<!doctype html>
<html lang="en">
<link rel="icon" type="image/png" href="/revbayes-site/assets/img/favicon.png" >
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="https://revbayes.github.io/revbayes-site/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/revbayes-site/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/revbayes-site/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/revbayes-site/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/revbayes-site/assets/css/main.css" />
    <title>RevBayes: Branch-Specific Diversification Rate Estimation</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <a href="/revbayes-site/" class="pull-left">
        
        <img class="navbar-logo" src="/revbayes-site/assets/img/aquabayes-desaturated.png" alt="RevBayes Home" />
        
      </a>
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar" align="right"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
        <li><a href="/revbayes-site/software">Software</a></li>
        <li><a href="/revbayes-site/tutorials/">Tutorials</a></li>
        <li><a href="/revbayes-site/workshops/">Workshops</a></li>
        <li><a href="/revbayes-site/developer/">Developer</a></li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <div class="titlebar">
	<h1 class="maintitle">Branch-Specific Diversification Rate Estimation</h1>
	<h3 class="subtitle">How to estimate branch-specific shifts in diversification rates</h3>
	<h4 class="authors">Sebastian H&#246;hna and Michael R. May</h4>
</div>


<div class="sidebar no-print">
<blockquote class="overview" id="overview">
  <h2>Overview</h2>
  
  <div class="row">
    <div class="col-md-9">
        <strong>Prerequisites</strong>
        
          <ul id="prerequisites">
          
            <li><a href="/revbayes-site/tutorials/intro/">Rev Language Syntax</a></li>
          
          </ul>
        
    </div>
  </div>
  
</blockquote>





<blockquote class="tutorial_files" id="tutorial_files">
    <h2>Data files and scripts</h2>
    
        
        <strong>Data Files</strong>
        <ul id="data_files">
        
        
          <li><a href="/revbayes-site/tutorials/branch_specific_div/data/primates_tree.nex">primates_tree.nex</a></li>
        
        </ul>
    
        
        <strong>Scripts</strong>
        <ul id="scripts">
        
        
          <li><a href="/revbayes-site/tutorials/branch_specific_div/scripts/mcmc_BSBD.Rev">mcmc_BSBD.Rev</a></li>
        
          <li><a href="/revbayes-site/tutorials/branch_specific_div/scripts/mcmc_CBDSP.Rev">mcmc_CBDSP.Rev</a></li>
        
          <li><a href="/revbayes-site/tutorials/branch_specific_div/scripts/mcmc_MRBD.Rev">mcmc_MRBD.Rev</a></li>
        
          <li><a href="/revbayes-site/tutorials/branch_specific_div/scripts/ml_MRBD.Rev">ml_MRBD.Rev</a></li>
        
        </ul>
    
</blockquote>


</div>


<h2>Required Software</h2>

For our tutorials we recommend that you download and install the latest release
of ‘RevBayes‘ <a href="#Hoehna2016b">(Höhna et al. 2016)</a>, which is available for Mac OS X, Windows,
and Linux operating systems. Directions for downloading and installing
the software are available on the program webpage:
<a href="http://revbayes.com">http://revbayes.com</a>. The exercises provided often also
require additional programs for editing text files and visualizing
output. The following are very useful tools for working with ‘RevBayes‘:

<ul>
<li>Some of the output can be automatically visualized using pre-made 
    <a href="https://www.r-project.org">R</a> functions provided by 
    our R-package <a href="https://github.com/revbayes/RevGadgets">RevGadgets</a>. 
    You should make sure that you have a recent release of 
    <a href="https://www.r-project.org">R</a> and 
    <a href="https://github.com/revbayes/RevGadgets">RevGadgets</a> installed.
</li>
<li>A good text editor – if you do not already have one that you like,
    we recommend one that has features for syntax coloring, easy
    navigation between different files, line numbers, etc. Good options
    include <a href="http://www.sublimetext.com/">Sublime Text</a>, 
    <a href="https://atom.io/">Atom</a> or <a href="https://notepad-plus-plus.org">NotePad++</a>, which are available for Mac OSX, Windows,
    and Linux.
</li>
<li><a href="http://tree.bio.ed.ac.uk/software/tracer/">Tracer</a> – for
    visualizing and assessing numerical parameter samples from
    ‘RevBayes‘
</li>
<li><a href="http://tgvaughan.github.io/icytree/">IcyTree</a> – a web-hosted
    phylogenetic tree visualization tool that is supported for
    <a href="https://www.mozilla.org/en-US/firefox/products/">Firefox</a> or
    <a href="https://www.google.com/chrome/">Google Chrome</a> browsers

<li><a href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a> – a tree
    visualization program
</li>
</ul>



<h1 id="sec:diversification_rate_overview">Overview: Diversification Rate Estimation</h1>

<p>Models of speciation and extinction are fundamental to any phylogenetic
analysis of macroevolutionary processes
(<em>e.g.</em>, divergence time estimation,
diversification rate estimation, continuous and discrete trait
evolution, and historical biogeography). First, a prior model describing
the distribution of speciation events over time is critical to
estimating phylogenies with branch lengths proportional to time. Second,
stochastic branching models allow for inference of speciation and
extinction rates. These inferences allow us to investigate key questions
in evolutionary biology.</p>

<p>Diversification-rate parameters may be included as nuisance parameters
of other phylogenetic models—<em>i.e.</em>, where
these diversification-rate parameters are not of direct interest. For
example, many methods for estimating species divergence times—such as
<code class="highlighter-rouge">BEAST</code> <a href="#Drummond2012">(Drummond et al. 2012)</a>,
<code class="highlighter-rouge">MrBayes</code> <a href="#Ronquist2012">(Ronquist et al. 2012)</a>, and RevBayes
<a href="#Hoehna2016b">(Höhna et al. 2016)</a>—implement <code class="highlighter-rouge">relaxed-clock models' that include a
constant-rate birth-death branching process as a prior model on the
distribution of tree topologies and node ages. Although the parameters
of these </code>tree priors’ are not typically of direct interest, they are
nevertheless estimated as part of the joint posterior probability
distribution of the relaxed-clock model, and so can be estimated simply
by querying the corresponding marginal posterior probability densities.
In fact, this may provide more robust estimates of the
diversification-rate parameters, as they accommodate uncertainty in the
other phylogenetic-model parameters (including the tree topology,
divergence-time estimates, and the other relaxed-clock model
parameters). More recent work,
<em>e.g.</em>, <a href="#Heath2014">(Heath et al. 2014)</a>, uses macroevolutionary
models (the fossilized birth-death process) to calibrate phylogenies and
thus to infer dated trees.</p>

<p>In these tutorials we focus on the different types of macroevolutionary
models to study diversification processes and thus the
diversification-rate parameters themselves. Nevertheless, these
macroevolutionary models should be used for other evolutionary
questions, when an appropriate prior distribution on the tree and
divergence times is needed.</p>

<h2 id="types-of-hypotheses-for-estimating-diversification-rates">Types of Hypotheses for Estimating Diversification Rates</h2>

<p>Many evolutionary phenomena entail differential rates of diversification
(speciation – extinction); <em>e.g.</em>, adaptive
radiation, diversity-dependent diversification, key innovations, and
mass extinction. The specific study questions regarding lineage
diversification may be classified within three fundamental categories of
inference problems. Admittedly, this classification scheme is somewhat
arbitrary, but it is nevertheless useful, as it allows users to navigate
the ever-increasing number of available phylogenetic methods. Below, we
describe each of the fundamental questions regarding diversification
rates.</p>

<h4 id="1-diversification-rate-through-time-estimation">(1) Diversification-rate through time estimation</h4>

<p><em>What is the (constant) rate of diversification in my study group?</em> The
most basic models estimate parameters of the stochastic-branching
process (<em>i.e.</em>, rates of speciation and
extinction, or composite parameters such as net-diversification and
relative-extinction rates) under the assumption that rates have remained
constant across lineages and through time;
<em>i.e.</em>, under a constant-rate birth-death
stochastic-branching process model <a href="#Nee1994b">(Nee et al. 1994)</a>. Extensions to the
(basic) constant-rate models include diversification-rate variation
through time (missing reference). First, we might ask whether
there is evidence of an episodic, tree-wide increase in diversification
rates (associated with a sudden increase in speciation rate and/or
decrease in extinction rate), as might occur during an episode of
adaptive radiation. A second question asks whether there is evidence of
a continuous/gradual decrease in diversification rates through time
(associated with decreasing speciation rates and/or increasing
extinction rates), as might occur because of diversity-dependent
diversification (<em>i.e.</em>, where competitive
ecological interactions among the species of a growing tree decrease the
opportunities for speciation and/or increase the probability of
extinction, <em>e.g.</em>, <a href="#Hoehna2014a">(Höhna 2014)</a>). Third, we
can ask whether changes in diversification rates are correlated with
environmental factors, such as environmental CO~2~ or temperature
<a href="#Condamine2013">(Condamine et al. 2013)</a>. A final question in this category asks whether our
study tree was impacted by a mass-extinction event (where a large
fraction of the standing species diversity is suddenly lost,
<em>e.g.</em>, <a href="#May2016">(May et al. 2016)</a>). The common theme of these
studies is that the diversification process is tree-wide, that is, all
lineages of the study group have the exact same rates at a given time.</p>

<h4 id="2-diversification-rate-variation-across-branches-estimation">(2) Diversification-rate variation across branches estimation</h4>

<p><em>Is there evidence that diversification rates have varied significantly
across the branches of my study group?</em> Models have been developed to
detect departures from rate constancy across lineages; these tests are
analogous to methods that test for departures from a molecular
clock—<em>i.e.</em>, to assess whether substitution
rates vary significantly across lineages (missing reference).
These models are important for assessing whether a given tree violates
the assumptions of rate homogeneity among lineages. Furthermore, these
models are important to answer questions such as: <em>What are the
branch-specific diversification rates?</em>; and <em>Have there been
significant diversification-rate shifts along branches in my study
group, and if so, how many shifts, what magnitude of rate-shifts and
along which branches?</em></p>

<h4 id="3-character-dependent-diversification-rate-estimation">(3) Character-dependent diversification-rate estimation</h4>

<p><em>Are diversification rates correlated with some variable in my study
group?</em> Character-dependent diversification-rate models aim to identify
overall correlations between diversification rates and organismal
features (binary and multi-state discrete morphological traits,
continuous morphological traits, geographic range, etc.). For example,
one can hypothesize that a binary character, say if an organism is
herbivorous/carnivorous or self-compatible/self-incompatible, impact the
diversification rates. Then, if the organism is in state 0
(<em>e.g.</em>, is herbivorous) it has a lower (or
higher) diversification rate than if the organism is in state 1
(<em>e.g.</em>, carnivorous) <a href="#Maddison2007">(Maddison et al. 2007)</a>.</p>

<h1 id="sec:models">Diversification Rate Models</h1>

<p>We begin this section with a general introduction to the stochastic
birth-death branching process that underlies inference of
diversification rates in RevBayes. This primer will provide some
details on the relevant theory of stochastic-branching process models.
We appreciate that some readers may want to skip this somewhat technical
primer; however, we believe that a better understanding of the relevant
theory provides a foundation for performing better inferences. We then
discuss a variety of specific birth-death models, but emphasize that
these examples represent only a tiny fraction of the possible
diversification-rate models that can be specified in RevBayes.</p>

<h2 id="the-birth-death-branching-process">The birth-death branching process</h2>

<p>Our approach is based on the <em>reconstructed evolutionary process</em>
described by <a href="#Nee1994b">(Nee et al. 1994)</a>; a birth-death process in which only sampled,
extant lineages are observed. Let $N(t)$ denote the number of species at
time $t$. Assume the process starts at time $t_1$ (the `crown’ age of
the most recent common ancestor of the study group, $t_\text{MRCA}$)
when there are two species. Thus, the process is initiated with two
species, $N(t_1) = 2$. We condition the process on sampling at least one
descendant from each of these initial two lineages; otherwise $t_1$
would not correspond to the $t_\text{MRCA}$ of our study group. Each
lineage evolves independently of all other lineages, giving rise to
exactly one new lineage with rate $b(t)$ and losing one existing lineage
with rate $d(t)$ (Figure [fig:BirthDeathShift] and
Figure [fig:BDP]). Note that although each lineage evolves
independently, all lineages share both a common (tree-wide) speciation
rate $b(t)$ and a common extinction rate $d(t)$
(missing reference). Additionally, at certain times,
$t_{\mathbb{M}}$, a mass-extinction event occurs and each species
existing at that time has the same probability, $\rho$, of survival.
Finally, all extinct lineages are pruned and only the reconstructed tree
remains (Figure [fig:BirthDeathShift]).</p>

<p>![A realization of the birth-death process with mass extinction.
Lineages that have no extant or sampled descendant are shown in gray and
surviving lineages are shown in a thicker black line.<span data-label="fig:BirthDeathShift">](\ResourcePath figures/BirthDeathShift.pdf){width=”80.00000%”}</span></p>

<blockquote class="figure">
  <p><img src="figures/BirthDeathShift.png" alt="" /> 
A realization of the
birth-death process with mass extinction. Lineages that have no extant
or sampled descendant are shown in gray and surviving lineages are shown
in a thicker black line.</p>
</blockquote>

<p>![ The process is initiated at the first speciation event (the
`crown-age’ of the MRCA) when there are two initial lineages. At each
speciation event the ancestral lineage is replaced by two descendant
lineages. At an extinction event one lineage simply terminates. (A) A
complete tree including extinct lineages. (B) The reconstructed tree of
tree from A with extinct lineages pruned away. (C) A <em>uniform</em> subsample
of the tree from B, where each species was sampled with equal
probability, $\rho$. (D) A <em>diversified</em> subsample of the tree from B,
where the species were selected so as to maximize diversity.<span data-label="fig:BDP">](\ResourcePath figures/birth-death-sketch.pdf){width=”\textwidth”}</span></p>

<blockquote class="figure">
  <p><img src="figures/birth-death-sketch.png" alt="" /> 
<strong>Examples of
trees produced under a birth-death process.</strong> The process is
initiated at the first speciation event (the `crown-age’ of the MRCA)
when there are two initial lineages. At each speciation event the
ancestral lineage is replaced by two descendant lineages. At an
extinction event one lineage simply terminates. (A) A complete tree
including extinct lineages. (B) The reconstructed tree of tree from A
with extinct lineages pruned away. (C) A <em>uniform</em> subsample of the tree
from B, where each species was sampled with equal probability, $\rho$.
(D) A <em>diversified</em> subsample of the tree from B, where the species were
selected so as to maximize diversity.</p>
</blockquote>

<p>To condition the probability of observing the branching times on the
survival of both lineages that descend from the root, we divide by
$P(N(T) &gt; 0 | N(0) = 1)^2$. Then, the probability density of the
branching times, $\mathbb{T}$, becomes <script type="math/tex">\begin{aligned}
P(\mathbb{T}) = \fra\frac{}{b}ace{P(N(T) = 1 \mid N(0) = 1)^2}^{\text{both initial lineages have one descendant}}}{ \underbrace{P(N(T) > 0 \mid N(0) = 1)^2}_{\text{both initial lineages survive}} } \times \prod_{i=2}^{n-1\frac{ }{b}ace{i \times b(t_i)}^{\text{speciation rate}} \time\frac{ }{b}ace{P(N(T) = 1 \mid N(t_i) = 1)}^\text{lineage has one descendant},\end{aligned}</script>
and the probability density of the reconstructed tree (topology and
branching times) is then <script type="math/tex">% <![CDATA[
\begin{aligned}
P(\Psi) = \; & \frac{2^{n-1}}{n!(n-1)!} \times \left( \frac{P(N(T) = 1 \mid N(0) = 1)}{P(N(T) > 0 \mid N(0) = 1)} \right)^2 \nonumber\\
		  \; & \times \prod_{i=2}^{n-1} i \times b(t_i) \times P(N(T) = 1 \mid N(t_i) = 1)
	\label{eq:tree_probability}\end{aligned} %]]></script></p>

<p>We can expand Equation ([eq:tree_probability]) by substituting
$P(N(T) &gt; 0 \mid N(t) =1)^2 \exp(r(t,T))$ for
$P(N(T) = 1 \mid N(t) = 1)$, where $r(u,v) = \int^v_u d(t)-b(t)dt$; the
above equation becomes <script type="math/tex">% <![CDATA[
\begin{aligned}
P(\Psi) = \; & \frac{2^{n-1}}{n!(n-1)!} \times \left( \frac{P(N(T) > 0 \mid N(0) =1 )^2 \exp(r(0,T))}{P(N(T) > 0 \mid N(0) = 1)} \right)^2 \nonumber\\
		  \; & \times \prod_{i=2}^{n-1} i \times b(t_i) \times P(N(T) > 0 \mid N(t_i) = 1)^2 \exp(r(t_i,T)) \nonumber\\
		= \; & \frac{2^{n-1}}{n!} \times \Big(P(N(T) > 0 \mid N(0) =1 ) \exp(r(0,T))\Big)^2 \nonumber\\
		  \; & \times \prod_{i=2}^{n-1} b(t_i) \times P(N(T) > 0 \mid N(t_i) = 1)^2 \exp(r(t_i,T)).
		\label{eq:tree_probability_substitution}\end{aligned} %]]></script> For a detailed
description of this substitution, see <a href="#Hoehna2015a">(Höhna 2015)</a>. Additional
information regarding the underlying birth-death process can be found in
@Thompson1975 [Equation 3.4.6] and <a href="#Nee1994b">(Nee et al. 1994)</a> for constant rates and
(missing reference) for arbitrary rate functions.</p>

<p>To compute the equation above we need to know the rate function,
$r(t,s) = \int_t^s d(x)-b(x) dx$, and the probability of survival,
$P(N(T)!&gt;!0|N(t)!=!1)$. <a href="#Yule1925">(Yule 1925)</a> and later <a href="#Kendall1948">(Kendall 1948)</a> derived
the probability that a process survives ($N(T) &gt; 0$) and the probability
of obtaining exactly $n$ species at time $T$ ($N(T) = n$) when the
process started at time $t$ with one species. Kendall’s results were
summarized in Equation (3) and Equation (24) in <a href="#Nee1994b">(Nee et al. 1994)</a>
<script type="math/tex">% <![CDATA[
\begin{aligned}
P(N(T)\!>\!0|N(t)\!=\!1) & = & \left(1+\int\limits_t^{T} \bigg(\mu(s) \exp(r(t,s))\bigg) ds\right)^{-1} \label{eq:survival} \\ \nonumber \\
P(N(T)\!=\!n|N(t)\!=\!1) & = & (1-P(N(T)\!>\!0|N(t)\!=\!1)\exp(r(t,T)))^{n-1} \nonumber\\
& & \times P(N(T)\!>\!0|N(t)\!=\!1)^2 \exp(r(t,T)) \label{eq:N} %]]></script> An
overview for different diversification models is given in
<a href="#Hoehna2015a">(Höhna 2015)</a>.</p>

<p><strong><em>Sidebar: Phylogenetic trees as observations</em></strong></p>

<p>The branching processes used here describe probability distributions on
phylogenetic trees. This probability distribution can be used to infer
diversification rates given an “observed” phylogenetic tree. In reality
we never observe a phylogenetic tree itself. Instead, phylogenetic trees
themselves are estimated from actual observations, such as DNA
sequences. These phylogenetic tree estimates, especially the divergence
times, can have considerable uncertainty associated with them. Thus, the
correct approach for estimating diversification rates is to include the
uncertainty in the phylogeny by, for example, jointly estimating the
phylogeny and diversification rates. For the simplicity of the following
tutorials, we take a shortcut and assume that we know the phylogeny
without error. For publication quality analysis you should always
estimate the diversification rates jointly with the phylogeny and
divergence times.</p>

<h1 id="estimating-branch-specific-speciation--extinction-rates">Estimating Branch-Specific Speciation &amp; Extinction Rates</h1>

<h2 id="outline">Outline</h2>

<p>This tutorial describes how to specify a branch-specific
branching-process models in RevBayes; a birth-death process where
diversification rates vary among branches, similar to <a href="#Rabosky2014a">(Rabosky 2014)</a>.
The probabilistic graphical model is given for each component of this
tutorial. The goal is to obtain estimate of branch-specific
diversification rates using Markov chain Monte Carlo (MCMC).</p>

<h2 id="requirements">Requirements</h2>

<p>We assume that you have read and hopefully completed the following
tutorials:</p>

<ul>
  <li>
    <p>RB_Getting_Started</p>
  </li>
  <li>
    <p>RB_Basics_Tutorial</p>
  </li>
  <li>
    <p>RB_BayesFactor_Tutorial</p>
  </li>
  <li>
    <p>RB_BasicDiversificationRate_Tutorial</p>
  </li>
</ul>

<p>Note that the RB_Basics_Tutorial introduces the basic syntax of <code class="highlighter-rouge">Rev</code>
but does not cover any phylogenetic models. You may skip the
RB_Basics_Tutorial if you have some familiarity with <code class="highlighter-rouge">R</code>.
The RB_BayesFactor_Tutorial introduces Bayesian model selection by
means of Bayes factors, which can be skipped by readers familiar with
Bayesian model selection. We tried to keep this tutorial very basic and
introduce all the language concepts and theory on the way. You may only
need the RB_Basics_Tutorial for a more in-depth discussion of concepts
in <code class="highlighter-rouge">Rev</code>.</p>

<h1 id="data-and-files">Data and files</h1>

<p>We provide the data file(s) which we will use in this tutorial. You may
want to use your own data instead. In the <code class="highlighter-rouge">data</code> folder, you will find
the following files</p>

<ul>
  <li><code class="highlighter-rouge">primates_tree.nex</code>: Dated primates phylogeny including 233 out of
367 species from <a href="#MagnusonFord2012">(Magnuson-Ford and Otto 2012)</a>.</li>
</ul>

<p>Open the tree <code class="highlighter-rouge">data/primates_tree.nex</code> in FigTree.</p>

<h1 id="branch-specific-birth-death-model">Branch-Specific Birth-Death Model</h1>

<blockquote class="figure">
  <p><img src="figures/stochastic_process_figure.png" alt="" /> 
Cartoon of a
branch-specific birth-death process. On the left we see the full
process. On the right we only see the branches of the reconstructed
tree, thus missing one rate-shift event.</p>
</blockquote>

<blockquote class="figure">
  <p><img src="figures/stochastic_process_figure.png" alt="" /> 
Cartoon of a
branch-specific birth-death process. On the left we see the full
process. On the right we only see the branches of the reconstructed
tree, thus missing one rate-shift event.</p>
</blockquote>

<p>The basic idea behind the model is that speciation and extinction rates
are allowed to vary across branches of the tree (see Figure
[fig:BSBD]). Unfortunately, it is not possible to model rates drawn
from a continuous distribution directly, as done for example in
<code class="highlighter-rouge">BAMM</code>, because in that case one needs to integrate over
any number of possible rate shifts, any time of these shifts and most
importantly over all possible new rates. This is unfeasible to do and
failure to do so has been shown to make parameter estimates unreliable
<a href="#Moore2016">(Moore et al. 2016)</a>.</p>

<blockquote class="figure">
  <p><img src="figures/likelihood_figure.png" alt="" /> 
Cartoon of the
likelihood computation using numerical integration.</p>
</blockquote>

<blockquote class="figure">
  <p><img src="figures/likelihood_figure.png" alt="" /> 
Cartoon of the
likelihood computation using numerical integration.</p>
</blockquote>

<p>Here we adopt an approach using (few) discrete rate categories instead.
This allows us to numerically integrate over all possible rate
categories using a system of differential equation originally described
by <a href="#Maddison2007">(Maddison et al. 2007)</a> (see also <a href="#FitzJohn2009">(FitzJohn et al. 2009)</a> and <a href="#FitzJohn2010">(FitzJohn 2010)</a>). The
numerical procedure beaks time into very small time intervals and sums
over all possible events occurring in that interval (see Figure
[fig:BSBD_likelihood]).</p>

<p>You don’t need to worry about any of the technical details. It is
important for you to realize that this model assumes that new rates at a
rate-shift event are drawn from a given (discrete) set of rates.</p>

<p>In RevBayes we have two implementations
(<em>i.e.</em>, distributions) for modeling a
branch-specific birth-death process. The first distribution is the
<code class="highlighter-rouge">dnBirthDeathMultiRate</code> (or its alias <code class="highlighter-rouge">dnMRBDP</code>) and the second is the
<code class="highlighter-rouge">dnHeterogeneousBirthDeath</code> (or its alias <code class="highlighter-rouge">dnHBDP</code>). We have designed
this tutorial so that each section can be read independently although we
recommend that you work through both of them. You will find that some
parts are redundant, which is intentional to emphasize the similarities
between the analysis but also to make the sections independent.</p>

<h1 id="testing-for-branch-specific-diversification-rates">Testing for Branch-Specific-Diversification Rates</h1>

<p>In this first exercise we are interested in knowing if there is
diversification-rate variation among branches for our study tree. That
is, we want to see if we can reject a constant rate birth-death process.
Therefore, we don’t focus on branch-specific parameter estimates but
instead on the marginal likelihood estimation for model testing.</p>

<p>We assume that you have completed the
RB_BasicDiversificationRate_Tutorial to estimate the marginal
likelihood under a constant-rate birth-death process. If you haven’t
done so, then you should go back and do this now!</p>

<h2 id="read-the-tree">Read the tree</h2>

<p>Begin by reading in the observed tree.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>observed_phylogeny &lt;- readTrees("data/primates_tree.nex")[1]
</code></pre></div></div>

<p>From this tree, we can get some helpful variables:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>taxa &lt;- observed_phylogeny.taxa()
root &lt;- observed_phylogeny.rootAge()
tree_length &lt;- observed_phylogeny.treeLength()
</code></pre></div></div>

<p>Additionally, we can initialize an iterator variable for our vector of
moves and monitors:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mvi = 0
mni = 0
</code></pre></div></div>

<p>Finally, we create a helper variable that specifies the number of
discrete rate categories, another helper variable for the expected
number of rate-shift events, the total number of species, and the
variation in rates.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>NUM_RATE_CATEGORIES = 4
EXPECTED_NUM_EVENTS = 2
NUM_TOTAL_SPECIES = 367
H = 0.587405
</code></pre></div></div>

<p>Using these variables we can easily change our script, for example, to
use more or fewer categories and test the impact. For example, setting
<code class="highlighter-rouge">NUM_RATE_CATEGORIES = 1</code> gives the constant rate birth-death process.</p>

<h2 id="specifying-the-model">Specifying the model</h2>

<h3 id="priors-on-rates">Priors on rates</h3>

<blockquote class="figure">
  <p><img src="figures/discretized_lognormal.png" alt="" /> 
Discretization of
a lognormal distribution. The two left figures have 4 rate categories
and the two right plots have 10 rate categories. The top plots have the
95% probability interval spanning one order of magnitude (<code class="highlighter-rouge">sd</code>
$=0.587405$) and the bottom plots have the 95% probability interval
spanning two orders of magnitude (<code class="highlighter-rouge">sd</code> $=2*0.587405$)</p>
</blockquote>

<blockquote class="figure">
  <p><img src="figures/discretized_lognormal.png" alt="" /> 
Discretization of
a lognormal distribution. The two left figures have 4 rate categories
and the two right plots have 10 rate categories. The top plots have the
95% probability interval spanning one order of magnitude (<code class="highlighter-rouge">sd</code>
$=0.587405$) and the bottom plots have the 95% probability interval
spanning two orders of magnitude (<code class="highlighter-rouge">sd</code> $=2*0.587405$)</p>
</blockquote>

<p>Instead of using a continuous probability distribution we will use a
discrete approximation of the distribution, as done for modeling rate
variation across sites <a href="#Yang1994a">(Yang 1994)</a> and for modeling relaxed molecular
clocks <a href="#Drummond2006">(Drummond et al. 2006)</a>. That means, we assume that the speciation rates
are drawn from one of the $N$ quantiles of the lognormal distribution.
For this we will use the function <code class="highlighter-rouge">fnDiscretizeDistribution</code> which takes
in a distribution as its first argument and the number of quantiles as
the second argument. The return value is a vector of quantiles. We use
it as a deterministic variable and every time the parameters of the base
distribution (<em>i.e.</em>, the lognormal
distribution in our case) change the quantiles will update automatically
as well. Thus we only need to specify parameters for our base
distribution, the lognormal distribution. We choose a stochastic
variable for the mean parameter of the lognormal distribution drawn from
yet another lognormal prior distribution. We fix the prior mean on this
mean speciation rate on our expected diversification rate, which is
$\ln( \ln(\frac{#Taxa}{2})/age )$. Remember that the median of a
lognormal distribution is equal to the exponential of the mean
parameter. This is why we used a log-transform of the actual mean. This
prior density is analogous to the prior on the speciation-rate parameter
in the constant-rate birth-death process.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>speciation_prior_mean &lt;- ln( ln(NUM_TOTAL_SPECIES/2.0) / root )
speciation_mean ~ dnLognormal(mean=speciation_prior_mean, sd=H)
moves[++mvi] = mvScale(speciation_mean,lambda=1,tune=true,weight=5)
</code></pre></div></div>

<p>Additionally, we choose a fixed standard deviation of $2<em>H$
($0.587405</em>2$) for the speciation rates because it represents two orders
of magnitude variance in the rate categories.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>speciation_sd &lt;- 2*H
speciation_categories := fnDiscretizeDistribution( dnLognormal(ln(speciation_mean), speciation_sd), NUM_RATE_CATEGORIES )
</code></pre></div></div>

<p>We also need discretized extinction-rate categories. We are completely
free to choose how we construct these rate categories. For example, we
could choose a similar discretization of a lognormal distribution using
its quantiles to provide different extinction-rate categories. For
simplicity, this is how we specify the current model. Alternatively, we
could assume that each rate category has the same extinction rate.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>extinction_prior_mean &lt;- ln( ln(NUM_TOTAL_SPECIES/2.0) / root )
extinction_mean ~ dnLognormal(mean=extinction_prior_mean,sd=2*H)
moves[++mvi] = mvScale(extinction_mean,lambda=1.0,tune=true,weight=3.0)
</code></pre></div></div>

<p>As with the speciation rate, we discretize the lognormal distribution
into a finite number of rate categories.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>extinction_categories := fnDiscretizeDistribution( dnLognormal(ln(extinction_mean), H), NUM_RATE_CATEGORIES )
</code></pre></div></div>

<p>Now, we must create a vector that contains each combination of
speciation- and extinction-rates. This allows the rate of speciation to
change without changing the rate of extinction and vice versa. The
resulting vector should be $N^2$ elements long. We call these the
`paired’ rate categories.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>k = 1
for(i in 1:NUM_RATE_CATEGORIES) {
    for(j in 1:NUM_RATE_CATEGORIES) {
        speciation[k]   := speciation_categories[i]
        extinction[k++] := extinction_categories[j]
    }
}
</code></pre></div></div>

<p>Next, we need a rate parameter for the rate-shifts events. We do not
have much prior information about this rate but we can provide some
realistic ranges. For example, we can specify a mean rate so that the
resulting number of expected rate-shift events is 2 (as specified in our
global variable <code class="highlighter-rouge">EXPECTED_NUM_EVENTS</code>). Furthermore, we can say that
the 95% prior ranges exactly one order of magnitude. We achieve all this
by specifying a lognormal prior distribution with mean <code class="highlighter-rouge">ln(
EXPECTED_NUM_EVENTS/tree_length )</code> and standard deviation of <code class="highlighter-rouge">H</code>.
Remember that this is only possible if the tree is known and not
estimated simultaneously because only if the tree is do we also know the
tree length. As usual for rate parameter, we apply a scaling move to the
<code class="highlighter-rouge">event_rate</code> variable.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>event_rate ~ dnLognormal( ln( EXPECTED_NUM_EVENTS/tree_length ), H)
moves[++mvi] = mvScale(event_rate,lambda=1,tune=true,weight=5)
</code></pre></div></div>

<p>Additionally, we need a rate-matrix parameter providing the relative
rates between paired rate categories. In this case we simply use equal
rates between each rate category; and thus use the Jukes-Cantor rate
matrix. You could, for example, also use an ordered rate matrix where
the process needs to go through rate 2 before going to rate 3 when
starting in rate 1.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>rate_matrix &lt;- fnJC( NUM_RATE_CATEGORIES * NUM_RATE_CATEGORIES )
</code></pre></div></div>

<p>Furthermore, we need prior probabilities for the process being in either
paired rate category at the root. Given our lack of prior knowledge we
create a flat prior distribution giving each rate category equal weight.
We do this by create a constant variable using the simplex function.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>rate_category_prior &lt;- simplex( rep(1, NUM_RATE_CATEGORIES * NUM_RATE_CATEGORIES) )
</code></pre></div></div>

<h3 id="incomplete-taxon-sampling">Incomplete Taxon Sampling</h3>

<p>We know that we have sampled 233 out of 367 living primate species. To
account for this we can set the sampling parameter as a constant node
with a value of 233 / 367.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>rho &lt;- observed_phylogeny.ntips() / NUM_TOTAL_SPECIES
</code></pre></div></div>

<h3 id="root-age">Root age</h3>

<p>The birth-death process requires a parameter for the root age. In this
exercise we use a fix tree and thus we know the age of the tree. Hence,
we can get the value for the root from the @MagnusonFord2012 tree. This
is done using our global variable <code class="highlighter-rouge">root</code> defined above and nothing else
has to be done here.</p>

<h3 id="the-time-tree">The time tree</h3>

<p>Now we have all of the parameters we need to specify the full episodic
birth-death model. We initialize the stochastic node representing the
time tree.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>timetree ~ dnMRBDP(lambda=speciation, mu=extinction, Q=rate_matrix, rootAge=root, rho=rho, pi=rate_category_prior, delta=event_rate, taxa=taxa)
</code></pre></div></div>

<p>And then we attach data to it.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>timetree.clamp(observed_phylogeny)
</code></pre></div></div>

<p>Finally, we create a workspace object of our whole model using the
<code class="highlighter-rouge">model()</code> function.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mymodel = model(speciation)
</code></pre></div></div>

<p>The <code class="highlighter-rouge">model()</code> function traversed all of the connections and found all of
the nodes we specified.</p>

<h2 id="running-a-marginal-likelihood-estimation">Running a marginal likelihood estimation</h2>

<h3 id="specifying-monitors">Specifying Monitors</h3>

<p>For the marginal likelihood analysis we don’t necessarily need monitors
because we are not going to look into the samples. However, as good
practice we still define our two standard monitors: the model monitor
and a screen monitor</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>monitors[++mni] = mnModel(filename="output/primates_MRBD.log",printgen=10, separator = TAB)
monitors[++mni] = mnScreen(printgen=10, diversification_mean, turnover)
</code></pre></div></div>

<h3 id="initializing-and-running-the-mcmc-simulation">Initializing and Running the MCMC Simulation</h3>

<p>If you don’t feel comfortable with Bayesian model selection
anymore, then have a look at the
<a href="https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_BayesFactor_Tutorial/RB_BayesFactor_Tutorial.pdf">RB_BayesFactor_Tutorial</a>
again.</p>

<p>First, we create the variable containing the power posterior. This
requires us to provide a model and vector of moves, as well as an output
file name. The <code class="highlighter-rouge">cats</code> argument sets the number of power steps.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pow_p = powerPosterior(mymodel, moves, "output/MRBD_powp.out", cats=100)
</code></pre></div></div>

<p>We can start the power posterior by first burning in the chain and and
discarding the first 5000 states.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pow_p.burnin(generations=5000,tuningInterval=200)
</code></pre></div></div>

<p>Now execute the run with the <code class="highlighter-rouge">.run()</code> function:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pow_p.run(generations=2000)
</code></pre></div></div>

<p>Once the power posteriors have been saved to file, create a
stepping-stone sampler. This function can read any file of power
posteriors and compute the marginal likelihood using stepping-stone
sampling.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ss = steppingStoneSampler(file="output/MRBD_powp.out", powerColumnName="power", likelihoodColumnName="likelihood")
</code></pre></div></div>

<p>Compute the marginal likelihood under stepping-stone sampling using the
member function <code class="highlighter-rouge">marginal()</code> of the <code class="highlighter-rouge">ss</code> variable and record the value
in Table [tab:ss].</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ss.marginal()
</code></pre></div></div>

<p>Path sampling is an alternative to stepping-stone sampling and also
takes the same power posteriors as input.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ps = pathSampler(file="output/MRBD_powp.out", powerColumnName="power", likelihoodColumnName="likelihood")
</code></pre></div></div>

<p>Compute the marginal likelihood under stepping-stone sampling using the
member function <code class="highlighter-rouge">marginal()</code> of the <code class="highlighter-rouge">ps</code> variable and record the value
in Table [tab:ss].</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ps.marginal()
</code></pre></div></div>

<p>The <code class="highlighter-rouge">Rev</code> file for performing this analysis:
<a href="https://github.com/revbayes/revbayes_tutorial/raw/master/RB_DiversificationRateBranchSpecific_Tutorial/RevBayes_scripts/ml_MRBD.Rev"><code class="highlighter-rouge">ml_MRBD.Rev</code></a>.</p>

<h2 id="exercise-1">Exercise 1</h2>

<ul>
  <li>
    <p>Enter the marginal likelihood estimate from the previous exercise on
the constant-rate birth-death process in the table below.</p>
  </li>
  <li>
    <p>Compute the marginal likelihood under the 2-rate model,
<em>i.e.</em>, set the NUM_Rate_CATEGORIES
variable to 2.</p>
  </li>
  <li>
    <p>Repeat the estimation of the marginal likelihoods with other number
of rate categories to fill out the table.</p>
  </li>
  <li>
    <p>What is the most supported model? Can we reject the constant-rate
birth-death process?</p>
  </li>
</ul>

<p>l c c c c &amp; &amp; &amp; &amp;\
Marginal likelihood constant-rate ($N=1$) &amp;</p>

<p>&amp; &amp;</p>

<p>&amp;\
Marginal likelihood two rate ($N=2$) &amp;</p>

<p>&amp; &amp;</p>

<p>&amp;\
Marginal likelihood four rate ($N=4$) &amp;</p>

<p>&amp; &amp;</p>

<p>&amp;\
Marginal likelihood six rate ($N=6$) &amp;</p>

<p>&amp; &amp;</p>

<p>&amp;\
Marginal likelihood eight rate ($N=8$) &amp;</p>

<p>&amp; &amp;</p>

<p>&amp;\
Marginal likelihood ten rate ($N=10$) &amp;</p>

<p>&amp; &amp;</p>

<p>&amp;\
Supported model? &amp;</p>

<p>&amp; &amp;</p>

<p>&amp;\
\</p>

<p>[tab:ss]</p>

<h1 id="estimating-branch-specific-diversification-rates">Estimating Branch-Specific Diversification Rates</h1>

<p>In this second analysis we are interested in estimating the
branch-specific diversification rates. We are going to use a very
similar model to the one described in the previous section. However, now
we are going to use the <code class="highlighter-rouge">dnHBDP</code> distribution instead which will require
some slightly different parameterization and moves. The main difference,
as mentioned above, is that the <code class="highlighter-rouge">dnHBDP</code> uses a data-augementation
scheme to sample the locations and parameters of rate-shift events
across branches of the tree.</p>

<h2 id="read-the-tree-1">Read the tree</h2>

<p>Begin by reading in the observed tree.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>observed_phylogeny &lt;- readTrees("data/primates_tree.nex")[1]
</code></pre></div></div>

<p>From this tree, we can get some helpful variables:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>taxa &lt;- observed_phylogeny.taxa()
root &lt;- observed_phylogeny.rootAge()
tree_length &lt;- observed_phylogeny.treeLength()
</code></pre></div></div>

<p>Additionally, we can initialize an iterator variable for our vector of
moves:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mvi = 0
mni = 0
</code></pre></div></div>

<p>Finally, we create a helper variable that specifies the number of
discrete rate categories, another helper variable for the expected
number of rate-shift events, the total number of species, and the
variation in rates.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>NUM_RATE_CATEGORIES = 4
EXPECTED_NUM_EVENTS = 2
NUM_TOTAL_SPECIES = 367
H = 0.587405
</code></pre></div></div>

<p>Using these variables we can easily change our script, for example, to
use more or fewer categories and test the impact.</p>

<h2 id="specifying-the-model-1">Specifying the model</h2>

<h3 id="priors-on-rates-1">Priors on rates</h3>

<p>Similar the previous section, we will set up the rate categories using
the exact same model and <code class="highlighter-rouge">Rev</code> syntax. Thus, we first create our
hyper-prior on the mean speciation rate, which is drawn from a lognormal
distribution.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>speciation_prior_mean &lt;- ln( ln(NUM_TOTAL_SPECIES/2.0) / root_age )
speciation_mean ~ dnLognormal(mean=speciation_prior_mean, sd=H)
moves[++mvi] = mvScale(speciation_mean,lambda=1,tune=true,weight=5)
</code></pre></div></div>

<p>Additionally, we choose a fixed standard deviation of $H * 2$ for the
speciation rates because it represents two orders of magnitude variance
in the rate categories.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>speciation_sd &lt;- H*2
speciation_categories := fnDiscretizeDistribution( dnLognormal(ln(speciation_mean), speciation_sd), NUM_RATE_CATEGORIES )
</code></pre></div></div>

<p>We define the prior on the extinction rate in the same way as we did for
the speciation rate, with the only difference that we allow for two
orders of magnitude of uncertainty.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>extinction_prior_mean &lt;- ln( ln(NUM_TOTAL_SPECIES/2.0) / root_age )
extinction_mean ~ dnLognormal(mean=extinction_prior_mean,sd=H*2)
moves[++mvi] = mvScale(extinction_mean,lambda=1.0,tune=true,weight=3.0)
</code></pre></div></div>

<p>As with the speciation rate, we discretize the lognormal distribution
into a finite number of rate categories.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>extinction_categories := fnDiscretizeDistribution( dnLognormal(ln(extinction_mean), H), NUM_RATE_CATEGORIES )
</code></pre></div></div>

<p>Now, we must create a vector that contains each combination of
speciation- and extinction-rates. This allows the rate of speciation to
change without changing the rate of extinction and vice versa. The
resulting vector should be $N^2$ elements long. We call these the
`paired’ rate categories.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>k = 1
for(i in 1:NUM_RATE_CATEGORIES) {
    for(j in 1:NUM_RATE_CATEGORIES) {
        speciation[k]   := speciation_categories[i]
        extinction[k++] := extinction_categories[j]
    }
}
</code></pre></div></div>

<p>Next, we need a rate parameter for the rate-shifts events. We do not
have much prior information about this rate but we can provide some
realistic ranges. For example, we can specify a mean rate so that the
resulting number of expected rate-shift events is 2 (as specified in our
global variable <code class="highlighter-rouge">EXPECTED_NUM_EVENTS</code>). Furthermore, we can say that
the 95% prior ranges exactly one order of magnitude. We achieve all this
by specifying a lognormal prior distribution with mean <code class="highlighter-rouge">ln(
EXPECTED_NUM_EVENTS/tree_length )</code> and standard deviation of <code class="highlighter-rouge">H</code>.
Remember that this is only possible if the tree is known and not
estimated simultaneously because only if the tree is do we also know the
tree length. As usual for rate parameter, we apply a scaling move to the
<code class="highlighter-rouge">event_rate</code> variable.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>event_rate ~ dnLognormal( ln( EXPECTED_NUM_EVENTS/tree_length ), H)
moves[++mvi] = mvScale(event_rate,lambda=1,tune=true,weight=5)
</code></pre></div></div>

<p>Additionally, we need a parameter for the category of the process at
root. We use a uniform prior distribution on the indices 1 to $N^2$
since we do not have any prior information in which rate category the
process is at the root. The move for this random variable is a random
integer walk because the random variable is defined only on the indices
(<em>i.e.</em>, with real number).</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>root_category ~ dnUniformNatural(1,NUM_RATE_CATEGORIES * NUM_RATE_CATEGORIES)
moves[++mvi] = mvRandomIntegerWalk(root_category,weight=1)
</code></pre></div></div>

<h3 id="incomplete-taxon-sampling-1">Incomplete Taxon Sampling</h3>

<p>We know that we have sampled 233 out of 367 living primate species. To
account for this we can set the sampling parameter as a constant node
with a value of 233 / 367.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>rho &lt;- observed_phylogeny.ntips() / NUM_TOTAL_SPECIES
</code></pre></div></div>

<h3 id="root-age-1">Root age</h3>

<p>The birth-death process requires a parameter for the root age. In this
exercise we use a fix tree and thus we know the age of the tree. Hence,
we can get the value for the root from the @MagnusonFord2012 tree. This
is done using our global variable <code class="highlighter-rouge">root</code> defined above and nothing else
has to be done here.</p>

<h3 id="the-time-tree-1">The time tree</h3>

<p>Now we have all of the parameters we need to specify the full
branch-specific birth-death model. We initialize the stochastic node
representing the time tree.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>timetree ~ dnHBDP(lambda=speciation, mu=extinction, rootAge=root, rho=rho, rootState=root_category, delta=event_rate, taxa=taxa )
</code></pre></div></div>

<p>And then we attach data to it.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>timetree.clamp(observed_phylogeny)
</code></pre></div></div>

<p>This specific implementation of the branch-specific birth-death process
augments the tree with rate-shift events. In order to sample the number,
the location, and the types of the rate-shift events, we have to apply
special moves to the tree. These moves will not change the tree but only
the augmented rate-shift events. We use a <code class="highlighter-rouge">mvBirthDeathEvent</code> to add and
remove events, a <code class="highlighter-rouge">mvEventTimeBeta</code> move to change the time and location
of the events, and a <code class="highlighter-rouge">mvDiscreteEventCategoryRandomWalk</code> to change the
the paired-rate category to which a rate-shift event belongs.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>moves[++mvi] = mvBirthDeathEvent(timetree,weight=2)
moves[++mvi] = mvEventTimeBeta(timetree,weight=2)
moves[++mvi] = mvDiscreteEventCategoryRandomWalk(timetree,weight=2)
</code></pre></div></div>

<p>In this analysis, we are interested in the branch-specific
diversification rates. So far we do not have any variables that directly
give us the number of rate-shift events per branch or the rates per
branch. Fortunately, we can construct deterministic variables and query
these properties from the tree. These function are made available by the
branch-specific birth-death process distribution.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>num_events := timetree.numberEvents()
avg_lambda := timetree.averageSpeciationRate()
avg_mu     := timetree.averageExtinctionRate()
avg_net    := avg_lambda - avg_mu
avg_rel    := avg_mu / avg_lambda

total_num_events := sum( num_events )
</code></pre></div></div>

<p>Finally, we create a workspace object of our whole model using the
<code class="highlighter-rouge">model()</code> function.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mymodel = model(speciation)
</code></pre></div></div>

<p>The <code class="highlighter-rouge">model()</code> function traversed all of the connections and found all of
the nodes we specified.</p>

<h2 id="running-an-mcmc-analysis">Running an MCMC analysis</h2>

<h3 id="specifying-monitors-1">Specifying Monitors</h3>

<p>For our MCMC analysis, we need to set up a vector of <em>monitors</em> to
record the states of our Markov chain. First, we will initialize the
model monitor using the <code class="highlighter-rouge">mnModel</code> function. This creates a new monitor
variable that will output the states for all model parameters when
passed into a MCMC function.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>monitors[++mni] = mnModel(filename="output/primates_BSBD.log",printgen=10, separator = TAB)
</code></pre></div></div>

<p>Additionally, we create an extended-Newick monitor. The extended-Newick
monitor writes the tree to a file and adds parameter values to the
branches and/or nodes of the tree. We can thus print the tree with the
average speciation and extinction rates, as well as the net
diversification (speciation - extinction) and relative extinction
(extinction / speciation) rates, for each branch into a file. We will
need this file later to estimate and visualize the posterior
distribution of the rates at the branches.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>monitors[++mni] = mnExtNewick(filename="output/primates_BSBD.trees", isNodeParameter=FALSE, printgen=10, separator = TAB, tree=timetree, avg_lambda, avg_mu, avg_net, avg_rel)
</code></pre></div></div>

<p>Finally, create a screen monitor that will report the states of
specified variables to the screen with <code class="highlighter-rouge">mnScreen</code>:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>monitors[++mni] = mnScreen(printgen=10, event_rate, mean_speciation, root_category, total_num_events)
</code></pre></div></div>

<h3 id="initializing-and-running-the-mcmc-simulation-1">Initializing and Running the MCMC Simulation</h3>

<p>With a fully specified model, a set of monitors, and a set of moves, we
can now set up the MCMC algorithm that will sample parameter values in
proportion to their posterior probability. The <code class="highlighter-rouge">mcmc()</code> function will
create our MCMC object:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mymcmc = mcmc(mymodel, monitors, moves)
</code></pre></div></div>

<p>First, we will run a pre-burnin to tune the moves and to obtain starting
values from the posterior distribution.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mymcmc.burnin(generations=1000,tuningInterval=200)
</code></pre></div></div>

<p>Now, run the MCMC:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mymcmc.run(generations=5000)
</code></pre></div></div>

<p>When the analysis is complete, you will have the monitored files in your
output directory. You can then visualize the branch-specific rates by
attaching them to the tree. This is actually done automatically in our
<code class="highlighter-rouge">mapTree</code> function.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>treetrace = readTreeTrace("output/primates_BSBD.trees", treetype="clock")
map_tree = mapTree(treetrace,"output/primates_BSBD_MAP.tree")
</code></pre></div></div>

<p>Now you can open the tree in <code class="highlighter-rouge">FigTree</code>.</p>

<p>The <code class="highlighter-rouge">Rev</code> file for performing this analysis:
<a href="https://github.com/revbayes/revbayes_tutorial/raw/master/RB_DiversificationRateBranchSpecific_Tutorial/RevBayes_scripts/mcmc_BSBD.Rev"><code class="highlighter-rouge">mcmc_BSBD.Rev</code></a>.</p>

<h2 id="exercise">Exercise</h2>

<ul>
  <li>
    <p>Run an MCMC simulation to estimate the posterior distribution of the
speciation rate and extinction rate.</p>
  </li>
  <li>
    <p>Visualize the branch-specific rates in <code class="highlighter-rouge">FigTree</code>.</p>
  </li>
  <li>
    <p>Do you see evidence for rate decreases or increases? What is the
general trend?</p>
  </li>
  <li>
    <p>Run the analysis using a different number of categories,
<em>e.g.</em>, 2 or 6. How do the rates change?</p>
  </li>
  <li>
    <p>Modify the model by specifying a prior on the log-diversification
and log-turnover rate and then estimate the diversification rates
through time. Do you see any differences in the estimates?</p>
  </li>
</ul>


<ol class="bibliography"><li><span id="Drummond2012">Drummond A.J., Suchard M.A., Xie D., Rambaut A. 2012. Bayesian phylogenetics with BEAUti and the BEAST 1.7. Molecular Biology and Evolution. 29:1969–1973.</span>

<a href="https://doi.org/10.1093/molbev/mss075">10.1093/molbev/mss075</a>

</li>
<li><span id="Ronquist2012">Ronquist F., Teslenko M., Mark P. van der, Ayres D.L., Darling A., Höhna S., Larget B., Liu L., Suchard M.A., Huelsenbeck J.P. 2012. MrBayes 3.2: efficient Bayesian phylogenetic inference and model choice across a large model space. Systematic Biology. 61:539–542.</span>

<a href="https://doi.org/10.1093/sysbio/sys029">10.1093/sysbio/sys029</a>

</li>
<li><span id="Hoehna2016b">Höhna S., Landis M.J., Heath T.A., Boussau B., Lartillot N., Moore B.R., Huelsenbeck J.P., Ronquist F. 2016. RevBayes: Bayesian Phylogenetic Inference Using Graphical Models and an Interactive Model-Specification Language. Systematic Biology. 65:726–736.</span>

<a href="https://doi.org/10.1093/sysbio/syw021">10.1093/sysbio/syw021</a>

</li>
<li><span id="Heath2014">Heath T.A., Huelsenbeck J.P., Stadler T. 2014. The fossilized birth-death process for coherent calibration of divergence-time estimates. Proceedings of the National Academy of Sciences. 111:E2957–E2966.</span>

<a href="https://doi.org/10.1073/pnas.1319091111">10.1073/pnas.1319091111</a>

</li>
<li><span id="Nee1994b">Nee S., May R.M., Harvey P.H. 1994. The Reconstructed Evolutionary Process. Philosophical Transactions: Biological Sciences. 344:305–311.</span>

<a href="https://doi.org/10.1098/rstb.1994.0068">10.1098/rstb.1994.0068</a>

</li>
<li><span id="Hoehna2014a">Höhna S. 2014. Likelihood Inference of Non-Constant Diversification Rates with Incomplete Taxon Sampling. PLoS One. 9:e84184.</span>

<a href="https://doi.org/10.1371/journal.pone.0084184">10.1371/journal.pone.0084184</a>

</li>
<li><span id="Condamine2013">Condamine F.L., Rolland J., Morlon H. 2013. Macroevolutionary perspectives to environmental change. Ecology Letters.</span>

<a href="https://doi.org/10.1111/ele.12062">10.1111/ele.12062</a>

</li>
<li><span id="May2016">May M.R., Höhna S., Moore B.R. 2016. A Bayesian Approach for Detecting the Impact of Mass-Extinction Events on Molecular Phylogenies When Rates of Lineage Diversification May Vary. Methods in Ecology and Evolution. 7:947–959.</span>

<a href="https://doi.org/10.1111/2041-210X.12563">10.1111/2041-210X.12563</a>

</li>
<li><span id="Maddison2007">Maddison W.P., Midford P.E., Otto S.P. 2007. Estimating a binary character’s effect on speciation and extinction. Systematic Biology. 56:701.</span>

<a href="https://doi.org/10.1080/10635150701607033">10.1080/10635150701607033</a>

</li>
<li><span id="Hoehna2015a">Höhna S. 2015. The time-dependent reconstructed evolutionary process with a key-role for mass-extinction events. Journal of Theoretical Biology. 380:321–331.</span>

<a href="https://doi.org/http://dx.doi.org/10.1016/j.jtbi.2015.06.005">http://dx.doi.org/10.1016/j.jtbi.2015.06.005</a>

</li>
<li><span id="Yule1925">Yule G.U. 1925. A mathematical theory of evolution, based on the conclusions of Dr. JC Willis, FRS. Philosophical Transactions of the Royal Society of London. Series B, Containing Papers of a Biological Character. 213:21–87.</span>

</li>
<li><span id="Kendall1948">Kendall D.G. 1948. On the Generalized "Birth-and-Death" Process. The Annals of Mathematical Statistics. 19:1–15.</span>

<a href="https://doi.org/10.1214/aoms/1177730285">10.1214/aoms/1177730285</a>

</li>
<li><span id="Rabosky2014a">Rabosky D.L. 2014. Automatic detection of key innovations, rate shifts, and diversity-dependence on phylogenetic trees. PLoS One. 9:e89543.</span>

<a href="https://doi.org/10.1371/journal.pone.0089543">10.1371/journal.pone.0089543</a>

</li>
<li><span id="MagnusonFord2012">Magnuson-Ford K., Otto S.P. 2012. Linking the Investigations of Character Evolution and Species Diversification. The American Naturalist. 180:225–245.</span>

<a href="https://doi.org/10.1086/666649">10.1086/666649</a>

</li>
<li><span id="Moore2016">Moore B.R., Höhna S., May M.R., Rannala B., Huelsenbeck J.P. 2016. Critically evaluating the theory and performance of Bayesian analysis of macroevolutionary mixtures. Proceedings of the National Academy of Sciences. 113:9569–9574.</span>

<a href="https://doi.org/10.1073/pnas.1518659113">10.1073/pnas.1518659113</a>

</li>
<li><span id="FitzJohn2009">FitzJohn R.G., Maddison W.P., Otto S.P. 2009. Estimating trait-dependent speciation and extinction rates from incompletely resolved phylogenies. Systematic Biology. 58:595–611.</span>

<a href="https://doi.org/10.1093/sysbio/syp067">10.1093/sysbio/syp067</a>

</li>
<li><span id="FitzJohn2010">FitzJohn R.G. 2010. Quantitative Traits and Diversification. Systematic Biology. 59:619–633.</span>

<a href="https://doi.org/10.1093/sysbio/syq053">10.1093/sysbio/syq053</a>

</li>
<li><span id="Yang1994a">Yang Z. 1994. Maximum Likelihood Phylogenetic Estimation from DNA Sequences with Variable Rates Over Sites: Approximate Methods. Journal of Molecular Evolution. 39:306–314.</span>

<a href="https://doi.org/10.1007/BF00160154">10.1007/BF00160154</a>

</li>
<li><span id="Drummond2006">Drummond A.J., Ho S.Y.W., Phillips M.J., Rambaut A. 2006. Relaxed Phylogenetics and Dating with Confidence. PLoS Biology. 4:e88.</span>

<a href="https://doi.org/10.1371/journal.pbio.0040088">10.1371/journal.pbio.0040088</a>

</li></ol>

<script type="text/javascript">
var _ol = document.querySelectorAll('ol');
for (var i = 0, elem_ol; elem_ol = _ol[i]; i++) {
	if ( elem_ol.classList == "bibliography" ) {
		var _li = elem_ol.getElementsByTagName("li");
		//for (var j = 0, elem_li; elem_li = _li[j]; j++)
		//{
		//	elem_li.innerHTML = elem_li.innerHTML.replace(/(https?:\/\/)([^\s<]+)/,"<a href=\"$1$2\">$2");
		//}
		if(_li.length > 0)
			elem_ol.outerHTML = "<h2>References</h2>"+elem_ol.outerHTML
	}
}
</script>

      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a> | <a href="/revbayes-site/license">License</a> | <a href="/revbayes-site/citation">Citation</a> | <a href="https://groups.google.com/forum/#!forum/revbayes-users">Users Forum</a>
    </div>
  </div>
  <br>
  </div>
</footer>

    </div>
    <script src="/revbayes-site/assets/js/vendor/jquery.min.js"></script>
<script src="/revbayes-site/assets/js/vendor/FileSaver.min.js"></script>
<script src="/revbayes-site/assets/js/vendor/jszip.min.js"></script>
<script src="/revbayes-site/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "Rev highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "Rev highlighter-rouge";
//   }
//   
// });
</script>

<script src="/revbayes-site/assets/js/base.js"></script>

<script type="text/javascript"
    src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
  });
</script>

  </body>
</html>
