<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="https://willpett.github.io/revbayes_tutorials/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/revbayes_tutorials/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/revbayes_tutorials/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/revbayes_tutorials/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/revbayes_tutorials/assets/css/main.css" />
    <title>RevBayes: Phylogenetic Inference using ‘RevBayes‘</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

      <a href="https://revbayes.github.io" class="pull-left">
        <img class="navbar-logo" src="/revbayes_tutorials/assets/img/aquabayes-desaturated.png" alt="RevBayes Logo" />
      </a>
      
      <a class="navbar-brand" href="/revbayes_tutorials/">Home</a>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
        <li><a href="/revbayes_tutorials/software/">Software</a></li>
        <li><a href="/revbayes_tutorials/developer/">Developer</a></li>
        <li><a href="/revbayes_tutorials/tutorials/">Tutorials</a></li>
        <li><a href="/revbayes_tutorials/workshops/">Workshops</a></li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <div class="row">
	<h1 class="maintitle">Phylogenetic Inference using ‘RevBayes‘</h1>
	<h3 class="subtitle">Discrete Morphology</h3>
	<h4 class="maintitle">April M. Wright and Michael J. Landis</h4>
</div>

      
<blockquote class="overview no-print" id="overview">
  <h2>Overview</h2>
  <div class="row">
    <div class="col-md-9">
        <strong>Prerequisites</strong>
        
          <ul id="prerequisites">
          
            
            
            <li><a href="/revbayes_tutorials/tutorials/intro/">Statistical Inference Using RevBayes</a></li>
          
          </ul>
        
    </div>
  </div>
</blockquote>

<blockquote class="tutorial_files no-print" id="tutorial_files">
  <h2>Data files and scripts</h2>
  
    <div class="row" id="script_row">
      <div class="col-md-9">
          <strong>Scripts</strong>
          <ul id="scripts"></ul>
        </div>
    </div>
</blockquote>
      <h1 id="introduction">Introduction</h1>

<p>While molecular data have become the default for building phylogenetic
trees for many types of evolutionary analysis, morphological data
remains important, particularly for analyses involving fossils. The use
of morphological data raises special considerations for model-based
methods for phylogenetic inference. Morphological data are typically
collected to maximize the number of parsimony-informative characters -
that is, the characters that provide information in favor of one
topology over another. Morphological characters also do not carry common
meanings from one character in a matrix to the next; character codings
are made arbitrarily. These two factors require extensions to our
existing phylogenetic models. Accounting for the complexity of
morphological characters remains challenging. This tutorial will provide
a discussion of modeling morphological characters, and will demonstrate
how to perform Bayesian phylogenetic analysis with morphology using
‘RevBayes‘ <a href="#Hoehna2016b">(Höhna et al. 2016)</a>.</p>

<h2 id="contents">Contents</h2>

<p>The Discrete Morphology guide contains several tutorials</p>

<ul>
  <li>
    <p>Section [sec:dm_overview]: Overview of the Discrete Morphological
models</p>
  </li>
  <li>
    <p>Section [sec:dm_simple]: A simple discrete morphology analysis</p>
  </li>
  <li>
    <p>Section [sec:dm_disc]: A model for allowing state frequency
variation across binary characters</p>
  </li>
  <li>
    <p>Section [sec:dm_dir]: A model for allowing state frequency
variation across binary and multistate characters</p>
  </li>
  <li>
    <p>Section [sec:trace]: Evaluating the MCMC</p>
  </li>
</ul>

<h2 id="recommended-tutorials">Recommended tutorials</h2>

<p>The Discrete Morphology tutorials assume the reader is familiar with the
content covered in the following ‘RevBayes‘ tutorials</p>

<ul>
  <li>
    <p><strong>‘Rev‘ Basics</strong></p>
  </li>
  <li>
    <p><strong>Molecular Models of Character Evolution</strong></p>
  </li>
  <li>
    <p><strong>Running and Diagnosing an MCMC Analysis</strong></p>
  </li>
  <li>
    <p><strong>Divergence Time Estimation and Node Calibrations</strong></p>
  </li>
</ul>

<h1 id="sec:dm_overview">Overview of Discrete Morphology Models</h1>

<p>[fig:module-gm]</p>

<p><img src="figures/tikz/Mk_model" alt="" /></p>
<blockquote>
  <p>Graphical model showing the Mk
model (left panel). Rev code specifying the Mk model is on the
right-hand panel.</p>
</blockquote>

<p>Molecular data forms the basis of most phylogenetic analyses today.
However, morphological characters remain relevant: Fossils often provide
our only direct observation of extinct biodiversity; DNA degradation can
make it difficult or impossible to obtain sufficient molecular data from
fragile museum specimens. Using morphological data can help researchers
include specimens in their phylogeny that might be left out of a
molecular tree.</p>

<p>To understand how morphological characters are modeled, it is important
to understand how characters are collected. Unlike in molecular data,
for which homology is algorithmically determined, homology in a
character is typically assessed by an expert. Biologists will typically
decide what characters are homologous by looking across specimens at the
same structure in multiple taxa; they may also look at the developmental
origin of structures in making this assessment <a href="#Phillips06">(Phillips 2006)</a>. Once
homology is determined, characters are broken down into states, or
different forms a single character can take. The state ‘0’ commonly
refers to absence, meaning that character is not present. In some
codings, absence will mean that character has not evolved in that group.
In others, absence means that that character has not evolved in that
group, and/or that that character has been lost in that group
<a href="#freudenstein05">(Freudenstein 2005)</a>. This type of coding is arbitrary, but both
<strong>non-random</strong> and <strong>meaningful</strong>, and poses challenges for how we model
the data.</p>

<p>Historically, most phylogenetic analyses using morphological characters
have been performed using the maximum parsimony optimality criterion.
Maximum parsimony analysis involves proposing trees from the
morphological data. Each tree is evaluated according to how many changes
it implied in the data, and the tree that requires the fewest changes is
preferred. In this way of estimating a tree, a character that does not
change, or changes only in one taxon, cannot be used to discriminate
between trees (i.e., it does not favor a topology). Therefore, workers
with parsimony typically do not collect characters that are parsimony
uninformative.</p>

<p>In 2001, Paul Lewis <a href="#Lewis2001">(Lewis 2001)</a> introduced a generalization of the
Jukes-Cantor model of sequence evolution for use with morphological
data. This model, called the Mk (Markov model, assuming each character
is in one of <em>k</em> states) model provided a mathematical formulation that
could be used to estimate trees from morphological data in both
likelihood and Bayesian frameworks. While this model is a useful step
forward, as a generalization of the Jukes-Cantor, it still makes fairly
simplistic assumptions. This tutorial will guide you through estimating
a phylogeny with the Mk model, and two useful extensions to the model.</p>

<h2 id="the-mk-model">The Mk Model</h2>

<p>The Mk model is a generalization of the Jukes-Cantor model of nucleotide
sequence evolution, which we discussed in <strong>Molecular Models of
Character Evolution</strong>. The Q matrix for a two-state Mk model looks like
so:</p>

<script type="math/tex; mode=display">% <![CDATA[
Q = \begin{pmatrix} -\mu_0 & \mu_{01} \\
\mu_{10} & -\mu_1  &\\
\end{pmatrix} \mbox{  ,} %]]></script>

<p>This matrix can be expanded to accommodate multi-state data, as well:</p>

<script type="math/tex; mode=display">% <![CDATA[
Q = \begin{pmatrix} -\mu_0 & \mu_{01} & \mu_{02} & \mu_{03} \\
\mu_{10} & -\mu_1  & \mu_{12} & \mu_{13} \\
\mu_{20} & \mu_{21} & -\mu_2  & \mu_{23} \\
\mu_{30} & \mu_{31} & \mu_{32} & -\mu_3 
\end{pmatrix} \mbox{  ,} %]]></script>

<p>However, the Mk model sets transitions to be equal from any state to any
other state. In that sense, our multistate matrix really looks like
this:</p>

<script type="math/tex; mode=display">% <![CDATA[
Q = \begin{pmatrix} -(k-1)\mu & \mu & \mu & \mu \\
\mu & -(k-1)\mu  & \mu & \mu \\
\mu & \mu & -(k-1)\mu  & \mu \\
\mu & \mu & \mu & -(k-1)\mu \\
\end{pmatrix} \mbox{  ,} %]]></script>

<p>Because this is a Jukes-Cantor-like model <a href="#Jukes1969">(Jukes and Cantor 1969)</a>, state
frequencies do not vary as a model parameter. These assumptions may seem
unrealistic. However, all models are a compromise between reality and
generalizability. Prior work has demonstrated that, in many conditions,
the model does perform adequately <a href="#Wright2014">(Wright and Hillis 2014)</a>. Because morphological
characters do not carry common meaning across sites in a matrix in the
way that nucleotide characters do, making assumptions that fit all
characters is challenging. A visualization of this simple model can be
seen in Fig. [fig:module-gm].</p>

<p>We will first perform a phylogenetic analysis using the Mk model. In
further sections, we will explore how to relax key assumptions of the Mk
model.</p>

<h2 id="ascertainment-bias-subsubascertainment-bias">Ascertainment Bias {#subsub:Ascertainment Bias}</h2>

<p>When Lewis first introduced the Mk model, he observed that branch
lengths on the trees were greatly inflated. The reason for this is that
when morphological characters are collected, characters that do not
vary, or vary in a non-parsimony-informative way (such as
autapomorphies) are excluded. Excluding these low-rate characters causes
the overall amount of evolution to be over-estimated. This causes an
inflation in the branch lengths <a href="#Lewis2001">(Lewis 2001)</a>.</p>

<p>Therefore, when performing a morphological phylogenetic analysis, it is
important to correct for this bias. There are numerous statistically
valid ways to perform this correction <a href="#allman08b">(Allman and Rhodes 2008)</a>. Original corrections
simulated invariant and non-parsimony informative characters along the
proposed tree. The likelihood of these characters would then be
calculated and used to normalize the total likelihood value. ‘RevBayes‘
implements a dynamic programming approach that calculates the same
likelihood, but does so faster.</p>

<h1 id="sec:dm_simple">Example: Inferring a Phylogeny of Fossil Bears Using the Mk Model</h1>

<p>In this example, we will use morphological character data from 18 taxa
of extinct bears <a href="#abella2011">(Abella et al. 2011)</a>. The dataset contains 62 binary
characters, a fairly typical dataset size for morphological characters.</p>

<h2 id="subsect:Exercise-Format">Tutorial Format</h2>

<p>This tutorial follows a specific format for issuing instructions and
information.</p>

<p>The boxed instructions guide you to complete tasks that are not part of
the ‘RevBayes‘ syntax, but rather direct you to create directories or
files or similar.</p>

<p>Information describing the commands and instructions will be written in
paragraph-form before or after they are issued.</p>

<p>All command-line text, including all ‘Rev‘ syntax, are given in
‘monotype font‘. Furthermore, blocks of ‘Rev‘ code that are needed to
build the model, specify the analysis, or execute the run are given in
separate shaded boxes. For example, we will instruct you to create a
constant node called ‘example‘ that is equal to ‘1.0‘ using the ‘&lt;-‘
operator like this:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>example &lt;- 1.0
</code></pre></div></div>

<p>It is important to be aware that some PDF viewers may render some
characters given as differently. Thus, if you copy and paste text from
this PDF, you may introduce some incorrect characters. Because of this,
we recommend that you type the instructions in this tutorial or copy
them from the scripts provided.</p>

<h2 id="subsect:Exercise-DataFiles">Data and Files</h2>

<p>On your own computer, create a directory called (or any name you like).</p>

<p>In this directory download and unzip the archive containing the data
files:
<a href="https://github.com/revbayes/revbayes_tutorial/tree/master/RB_DiscreteMorphology_Tutorial/data.zip">‘data.zip‘</a>.</p>

<p>This will create a folder called ‘data‘ that contains the files
necessary to complete this exercise.</p>

<h2 id="subsect:Exercise-GetStart">Getting Started</h2>

<p>Create a new directory (in ‘RB_DiscreteMorphology_Tutorial‘) called .
(If you do not have this folder, please refer to the directions in
section [subsect:Exercise-DataFiles].)</p>

<p>When you execute ‘RevBayes‘ in this exercise, you will do so within the
main directory you created (‘RB_DiscreteMorphology_Tutorial‘), thus,
if you are using a Unix-based operating system, we recommend that you
add the ‘RevBayes‘ binary to your path.</p>

<h2 id="subsect:Exercise-CreatingFiles">Creating ‘Rev‘ Files</h2>

<p>For complex models and analyses, it is best to create ‘Rev‘ script files
that will contain all of the model parameters, moves, and functions. In
this exercise, you will work primarily in your text editor<sup id="fnref:1"><a href="#fn:1" class="footnote">1</a></sup> and
create a set of modular files that will be easily managed and
interchanged. In this first section, you will write the following files
from scratch and save them in the ‘scripts‘ directory:</p>

<ul>
  <li>
    <p>‘mcmc_mk.Rev‘: the master ‘Rev‘ file that loads the data, the
separate model files, and specifies the monitors and MCMC sampler.</p>
  </li>
  <li>
    <p>‘model_mk.Rev‘: specifies the model describing discrete
morphological character change (binary characters).</p>
  </li>
</ul>

<p>All of the files that you will create are also provided in the
‘RevBayes‘ tutorial repository<sup id="fnref:2"><a href="#fn:2" class="footnote">2</a></sup>. Please refer to these files to
verify or troubleshoot your own scripts.</p>

<p>Open your text editor and create the master ‘Rev‘ file called in the
‘scripts‘ directory.</p>

<p>Enter the ‘Rev‘ code provided in this section in the new model file.</p>

<p>The file you will begin in this section will be the one you load into
‘RevBayes‘ when you’ve completed all of the components of the analysis.
In this section you will begin the file and write the ‘Rev‘ commands for
loading in the taxon list and managing the data matrices. Then, starting
in section [subsect:Exercise-MkModel], you will move on to writing
module files for each of the model components. Once the model files are
complete, you will return to editing ‘mcmc_Mk.Rev‘ and complete the
‘Rev‘ script with the instructions given in section
[subsect:Exercise-CompleteMCMC].</p>

<h3 id="subsub:Exercise-LoadData">Load Data Matrices</h3>

<p>‘RevBayes‘ uses the function ‘readDiscreteCharacterData()‘ to load a
data matrix to the workspace from a formatted file. This function can be
used for both molecular sequences and discrete morphological characters.
Import the morphological character matrix and assign it to the variable
‘morpho‘.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>morpho &lt;- readDiscreteCharacterData("data/bears.nex")
</code></pre></div></div>

<h3 id="subsub:Exercise-mviVar">Create Helper Variables</h3>

<p>Before we begin writing the ‘Rev‘ scripts for each of the model
components, we need to instantiate a couple “helper variables” that will
be used by downstream parts of our model specification files. These
variables will be used in more than one of the module files so it’s best
to initialize them in the master file.</p>

<p>Create a new constant node called ‘n_taxa‘ that is equal to the number
of species in our analysis (18). We will also create a constant node of
the taxon names. This list will be used to initialize the tree.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>taxa &lt;- morpho.names()
n_taxa &lt;- morpho.size() 
</code></pre></div></div>

<p>Next, create a workspace variable called ‘mvi‘. This variable is an
iterator that will build a vector containing all of the MCMC moves used
to propose new states for every stochastic node in the model graph. Each
time a new move is added to the vector, ‘mvi‘ will be incremented by a
value of ‘1‘.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mvi = 1
</code></pre></div></div>

<p>One important distinction here is that ‘mvi‘ is part of the ‘RevBayes‘
workspace and not the hierarchical model. Thus, we use the workspace
assignment operator ‘=‘ instead of the constant node assignment ‘&lt;-‘.</p>

<p>Save your current working version of ‘mcmc_Mk.Rev‘ in the ‘scripts‘
directory.</p>

<p>We will now move on to the next ‘Rev‘ file and will complete
‘mcmc_Mk.Rev‘ in section [subsect:Exercise-CompleteMCMC].</p>

<h2 id="subsect:Exercise-MkModel">The Mk Model</h2>

<p>Open your text editor and create the master ‘Rev‘ file called in the
‘scripts‘ directory.</p>

<p>Enter the ‘Rev‘ code provided in this section in the new model file.</p>

<p>First, we will create a vector of moves on branch lengths. This should
be familiar from the ‘RB_CTMC‘ tutorial:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>br_len_lambda ~ dnExp(0.2)
moves[mvi++] = mvScale(br_len_lambda, weight=5)
nbr &lt;- 2*names.size() - 3
for (i in 1:nbr){
    br_lens[i] ~ dnExponential(br_len_lambda)
    moves[mvi++] = mvScale(br_lens[i]) 
}
</code></pre></div></div>

<p>Next, we will create a Q matrix. Recall that the Mk model is simply a
generalization of the JC model. Therefore, we will create a 2x2 Q matrix
using ‘fnJC‘, which initializes Q-matrices with equal transition
probabilities between all states.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Q_morpho &lt;- fnJC(2)
</code></pre></div></div>

<p>Now that we have the basics of the model specified, we will add
Gamma-distributed rate variation and specify moves on the parameter to
the Gamma distribution.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>alpha_morpho ~ dnExponential( 1.0 )
rates_morpho := fnDiscretizeGamma( alpha_morpho, alpha_morpho, 4 )

#Moves on the parameters to the Gamma distribution.
moves[mvi++] = mvScale(alpha_morpho, lambda=0.01, weight=5.0)
moves[mvi++] = mvScale(alpha_morpho, lambda=0.1,  weight=3.0)
moves[mvi++] = mvScale(alpha_morpho, lambda=1,    weight=1.0)
</code></pre></div></div>

<p>Next we assemble the tree and specify a move on the topology.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>tau ~ dnUniformTopology(names)
phylogeny := treeAssembly(tau, br_lens)
moves[mvi++] = mvNNI(tau, weight=2*nbr)
moves[mvi++] = mvSPR(tau, weight=nbr)
tree_length := phylogeny.treeLength()
</code></pre></div></div>

<p>Lastly, we set up the CTMC. This should be familiar from the ‘RB_CTMC‘
tutorial. We see some familiar pieces: tree, Q matrix and site_rates.
We also have two new keywords: data type and coding. The data type
argument specifies the type of data - in our case, “Standard”, the
specification for morphology. Coding specifies what type of
ascertainment bias is expected. We are using the ‘variable’ correction,
as we have no invariant character in our matrix. If we also lacked
parsimony non-informative characters, we would use the coding
‘informative’.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>phyMorpho ~ dnPhyloCTMC(tree=phylogeny, siteRates=rates_morpho, Q=Q_morpho, type="Standard", coding="variable")
phyMorpho.clamp(morpho)
</code></pre></div></div>

<p>All of the components of the model are now specified.</p>

<h2 id="subsect:Exercise-CompleteMCMC">Complete Master ‘Rev‘ File</h2>

<p>Return to the master ‘Rev‘ file called in the ‘scripts‘ directory.</p>

<p>Enter the ‘Rev‘ code provided in this section in this file.</p>

<h3 id="subsub:Exercise-SourceMods">Source Model Scripts</h3>

<p>‘RevBayes‘ uses the ‘source()‘ function to load commands from ‘Rev‘
files into the workspace. Use this function to load in the model scripts
we have written in the text editor and saved in the ‘scripts‘ directory.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>source("scripts/model_Mk.Rev")
</code></pre></div></div>

<h3 id="subsub:Exercise-ModObj">Create Model Object</h3>

<p>We can now create our workspace model variable with our fully specified
model DAG. We will do this with the ‘model()‘ function and provide a
single node in the graph (‘phylogeny‘).</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mymodel = model(phylogeny)
</code></pre></div></div>

<p>The object ‘mymodel‘ is a wrapper around the entire model graph and
allows us to pass the model to various functions that are specific to
our MCMC analysis.</p>

<h3 id="subsub:Exercise-Monitors">Specify Monitors and Output Filenames</h3>

<p>The next important step for our master ‘Rev‘ file is to specify the
monitors and output file names. For this, we create a vector called
‘monitors‘ that will each sample and record or output our MCMC.</p>

<p>First, we will specify a workspace variable to iterate over the
‘monitors‘ vector.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mni = 1
</code></pre></div></div>

<p>The first monitor we will create will monitor every named random
variable in our model graph. This will include every stochastic and
deterministic node using the ‘mnModel‘ monitor. The only parameter that
is not included in the ‘mnModel‘ is the tree topology. Therefore, the
parameters in the file written by this monitor are all numerical
parameters written to a tab-separated text file that can be opened by
accessory programs for evaluating such parameters. We will also name the
output file for this monitor and indicate that we wish to sample our
MCMC every 10 cycles.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>monitors[mni++] = mnModel(filename="output/mk_simple.log", printgen=10)
</code></pre></div></div>

<p>The ‘mnFile‘ monitor writes any parameter we specify to file. Thus, if
we only cared about the branch lengths and nothing else (this is not a
typical or recommended attitude for an analysis this complex) we
wouldn’t use the ‘mnModel‘ monitor above and just use the ‘mnFile‘
monitor to write a smaller and simpler output file. Since the tree
topology is not included in the ‘mnModel‘ monitor (because it is not
numerical), we will use ‘mnFile‘ to write the tree to file by specifying
our ‘phylogeny‘ variable in the arguments.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>monitors[mni++] = mnFile(filename="output/mk_simple.trees", printgen=10, phylogeny)
</code></pre></div></div>

<p>The third monitor we will add to our analysis will print information to
the screen. Like with ‘mnFile‘ we must tell ‘mnScreen‘ which parameters
we’d like to see updated on the screen.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>monitors[mni++] = mnScreen(printgen=10)
</code></pre></div></div>

<p>Finally, we’ll create an ancestral state monitor, which is described in
more detail in Section [sec:anc_states].</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>monitors[mni++] = mnJointConditionalAncestralState(
    tree=phylogeny,
    ctmc=phyMorpho,
    filename="output/mk_simple.states.txt",
    type="Standard",
    printgen=10,
    withStartStates=false)
</code></pre></div></div>

<h3 id="set-up-the-mcmc">Set-Up the MCMC</h3>

<p>Once we have set up our model, moves, and monitors, we can now create
the workspace variable that defines our MCMC run. We do this using the
‘mcmc()‘ function that simply takes the three main analysis components
as arguments.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mymcmc = mcmc(mymodel, monitors, moves)
</code></pre></div></div>

<p>The MCMC object that we named ‘mymcmc‘ has a member method called
‘.run()‘. This will execute our analysis and we will set the chain
length to ‘10000‘ cycles using the ‘generations‘ option.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mymcmc.run(generations=20000)
</code></pre></div></div>

<p>Once our Markov chain has terminated, we will want ‘RevBayes‘ to close.
Tell the program to quit using the ‘q()‘ function.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>q()
</code></pre></div></div>

<p>You made it! Save all of your files.</p>

<h2 id="subsect:Exercise-RunMCMC">Execute the MCMC Analysis</h2>

<p>With all the parameters specified and all analysis components in place,
you are now ready to run your analysis. The ‘Rev‘ scripts you just
created will all be used by ‘RevBayes‘ and loaded in the appropriate
order.</p>

<p>Begin by running the ‘RevBayes‘ executable. In Unix systems, type the
following in your terminal (if the ‘RevBayes‘ binary is in your path):</p>

<p>Provided that you started ‘RevBayes‘ from the correct directory
(‘RB_DiscreteMorphology_Tutorial‘), you can then use the ‘source()‘
function to feed ‘RevBayes‘ your master script file (‘mcmc_mk.Rev‘).</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>source("scripts/mcmc_mk.Rev")
</code></pre></div></div>

<p>This will execute the analysis and you should see the following output
(though not the exact same values):</p>

<p>When the analysis is complete, ‘RevBayes‘ will quit and you will have a
new directory called ‘output‘ that will contain all of the files you
specified with the monitors (Sect. [subsub:Exercise-Monitors]).</p>

<h3 id="ascertainment-bias">Ascertainment Bias</h3>

<p>As discussed in (Sect. [subsub:Ascertainment Bias]), we also need to
correct for ascertainment bias. Once your initial Mk model estimation,
source the mcmc_simple.Rev script. This file is an estimation of the Mk
model without any correction for ascertainment bias. We will use this
for comparison in a moment.</p>

<h1 id="sec:dm_disc">Example: Relaxing the Assumption of Equal Transition Probabilities</h1>

<p>Make a copy of the MCMC and model files you made earlier. Call them
‘mcmc_mk_dicretized.Rev‘ and ‘model_mk_discretized.Rev‘. These will
contain the new model parameters and models.</p>

<p>[fig:module-gm]</p>

<p><img src="figures/tikz/morpho_gm" alt="" /></p>
<blockquote>
  <p>Graphical model demonstrating the
discretized Beta distribution for allowing variable state frequencies.</p>
</blockquote>

<p>The Mk model makes a number of assumptions, but one that may strike you
as unrealistic is the assumption that characters are equally likely to
change from any one state to any other state. That means that a trait is
as likely to be gained as lost. While this may hold true for some
traits, we expect that it may be untrue for many others.</p>

<p>‘RevBayes‘ has functionality to allow us to relax this assumption. We do
this by specifying a Beta prior on state frequencies. Remember from the
‘RB_CTMC‘ lesson that stationary frequencies impact how likely we are
to see changes in a character. For example, it may be very likely, in a
character, to change from 0 to 1. But if the frequency of 0 is very low,
we will still seldom see this change.</p>

<p>We can exploit the relationship between state frequencies and observed
changes to allow for variable Q matrices across characters (Fig. 2). To
do this, we generate a Beta distribution on state frequencies, and use
the state frequencies from that Beta distribution to generate a series
of Q-matrices to use to evaluate our data <a href="#Pagel2004">(Pagel and Meade 2004)</a>.</p>

<p>This type of model is called a <strong>mixture model</strong>. There are assumed to
be subdivisions in the data, which may require different parameters (in
this case, state frequencies). These subdivisions are not defined <em>a
priori</em>. This model has previously been shown to be effective for a
range of empirical and simulated datasets <a href="#Wright2016">(Wright et al. 2016)</a>.</p>

<h2 id="modifying-the-mcmc-file">Modifying the MCMC File</h2>

<p>At each place in which the output files are specified in the MCMC file,
change the output path so you don’t overwrite the output from the
previous exercise. For example, you might call your output file
‘output/mk_discretized.log‘ and ‘output/mk_discretized.trees‘. Change
source statement to indicate the new model file.</p>

<h2 id="modifying-the-model-file">Modifying the Model File</h2>

<p>Open the new model file that you created. We need to modify the way in
which the Q matrix is specified. We will use a discretized Beta
distribution to place a prior on state frequencies. The Beta
distribution has two parameters, $\alpha$ and $\beta$. These two
parameters specify the shape of the distribution. State frequencies will
be evaluated according to this distribution, in the same way that rate
variation is evaluated according to the Gamma distribution. The
discretized distribution is split into multiple classes, each with it’s
own set of frequencies for the 0 and 1 characters. The number of classes
can vary; we have chosen 4 for tractability.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>n_cats = 4
alpha_ofbeta ~ dnExponential( 1 )
beta_ofbeta ~ dnExponential( 1 )
moves[mvi++] = mvScale(alpha_ofbeta, lambda=1,    weight=1.0 )
moves[mvi++] = mvScale(alpha_ofbeta, lambda=0.1,  weight=3.0 )
moves[mvi++] = mvScale(alpha_ofbeta, lambda=0.01, weight=5.0 )
moves[mvi++] = mvScale(beta_ofbeta, lambda=1,    weight=1.0 )
moves[mvi++] = mvScale(beta_ofbeta, lambda=0.1,  weight=3.0 )
moves[mvi++] = mvScale(beta_ofbeta, lambda=0.01, weight=5.0 )
</code></pre></div></div>

<p>Above, we initialized the number of categories, the parameters to the
Beta distribution, and the moves on the parameters to the Beta.</p>

<p>Next, we set the categories to each represent a quadrant of the Beta
distribution specified by the ‘alpha_ofbeta‘ and ‘beta_ofbeta‘. The
+1 values are added to the beta shape and scale parameters
to prevent model overfitting.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cats := fnDiscretizeBeta(alpha_ofbeta+1, beta_ofbeta+1, 4)
</code></pre></div></div>

<p>If you were to print the ‘cats‘ variable, you would see a list of state
frequencies like so:</p>

<p>Using these state frequencies, we will generate a new vector of Q
matrices. Because we are varying the state frequencies, we must use a Q
matrix generation function that allows for state frequencies to vary as
a parameter. We will, therefore, use the ‘fnF81‘ function.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>for (i in 1:cats.size())
{
    Q[i] := fnF81(simplex(abs(1-cats[i]), cats[i]))
}
</code></pre></div></div>

<p>Once we’ve made the our vector of matrices, we specified moves on our
matrix vector:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>matrix_probs ~ dnDirichlet(v(1,1,1,1))
moves[mvi++] = mvSimplexElementScale(matrix_probs, alpha=10, weight=1.0) 
</code></pre></div></div>

<p>This Dirichlet prior says that no category is expected to have more
characters than another. If you expected some category to hold more of
the characters, you could put more weight on that category.</p>

<p>The only other specification that needs to change in the model file is
the CTMC:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>phyMorpho ~ dnPhyloCTMC(tree=phylogeny, siteRates=rates_morpho, Q=Q, type="Standard", coding="variable", siteMatrices=matrix_probs)
</code></pre></div></div>

<p>You’ll notice that we’ve added a command to tell the CTMC that we have
multiple site matrices that will be applied to different characters in
the matrix.</p>

<h3 id="set-up-the-mcmc-1">Set-Up the MCMC</h3>

<p>The MCMC chain set-up does not need to change. Run the new MCMC file,
just as you ran the plain Mk file. This estimation will take longer than
the Mk model, due to increased model complexity.</p>

<h1 id="sec:dm_dir">Site-Heterogeneous Discrete Morphology Model</h1>

<p>[fig:module-relaxed morphology]</p>

<p><img src="figures/tikz/morph_hyperprior" alt="" /></p>
<blockquote>
  <p>Graphical model
demonstrating the Dirichlet prior to allow variable state frequencies in
both binary and multistate data.</p>
</blockquote>

<p>In the previous example, we explored allowing among-character variation
in state frequencies. This is an excellent start for allowing more
complex models for morphology. But this approach also has several
shortcomings. First, because we use a Beta distribution, this model
really only works for binary data. Secondly, oftentimes, we will not
have a good idea of the shape of the distribution from which we expect
state frequencies to be drawn.</p>

<p>To accommodate for these concerns, ‘RevBayes‘ also has a model that is
similar to the CAT model <a href="#Lartillot2004">(Lartillot and Philippe 2004)</a>.</p>

<p>The site-heterogeneous discrete morphology model (SHDM) uses a
hyperprior on the prior on state frequencies to mix over different
possible combinations state frequencies. In this mixture model, F81
Q-matrices (an extension of the Jukes-Cantor which allows for different
state frequencies between characters) is initialized from a set of state
frequencies. The number of Q-matrices initialized is equal to the number
of user-defined categories, as in the discretized Beta model. The state
frequencies used to initialize the Q-matrices are drawn from a
Dirichelet prior distribution, which is generated by drawing values from
an exponential hyperprior distribution. This model is visualized in Fig.
3.</p>

<h2 id="example-site-heterogeneous-discrete-morphology-model">Example: Site-Heterogeneous Discrete Morphology Model</h2>

<p>Make a copy of the MCMC and model files you just made. Call them
‘mcmc_mk_hyperprior.Rev‘ and ‘model_mk_hyperprior.Rev‘. These will
contain the new model parameters and models.</p>

<h2 id="modifying-the-mcmc-file-1">Modifying the MCMC File</h2>

<p>At each place in which the output files are specified in the MCMC file,
change the output path so you don’t overwrite the output from the
previous exercise. For example, you might call your output file
‘output/mk_hyperprior.log‘ and ‘output/mk_hyperprior.trees‘. We will
also monitor Q_morpho and pi. Add Q_morpho and pi to the ‘mnScreen‘.
Change source statement to indicate the new model file.</p>

<h2 id="modifying-the-model-file-1">Modifying the Model File</h2>

<p>Open the new model file that you created. We need to modify the way in
which the Q-matrix is specified. First, we will create a hyperprior
called ‘dir_alpha‘ and specify a move on it.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>dir_alpha ~ dnExponential(1)
moves[mvi++] = mvScale(dir_alpha, lambda=1,    weight=1.0 )
moves[mvi++] = mvScale(dir_alpha, lambda=0.1,  weight=3.0 )
moves[mvi++] = mvScale(dir_alpha, lambda=0.01, weight=5.0 )
</code></pre></div></div>

<p>This hyperparameter, dir_alpha, will be used as a parameter to a
Dirichelet distribution from which our state frequencies will be drawn.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pi_prior := v(dir_alpha,dir_alpha)
</code></pre></div></div>

<p>If you were using multistate data, the dir_alpha can be repeated for
each state. Next, we will modify our previous loop to use these state
frequencies to initialize our Q-matrices.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>for(i in 1:n_cats)
{
	pi[i] ~ dnDirichlet(pi_prior)
    moves[mvi++] = mvSimplexElementScale(pi[i], alpha=10, weight=1.0) 
    
    Q_morpho[i] := fnF81(pi[i])
}
</code></pre></div></div>

<p>In the above loop, for each of our categories, we make a new draw of
state frequencies from our Dirichelet distribution (the shape of which
is determined by our dir_alpha values). We then use ‘fnF81‘ to make our
Q-matrices. For each ‘RevBayes‘ iteration, we will have 4 pi values and
4 Q-matrices, one for each of the number of categories we specified.</p>

<p>No other aspects of the model file need to change. Run the MCMC as
before.</p>

<h1 id="sec:trace">Evaluate and Summarize Your Results</h1>

<h2 id="subsub:Exercise-EvalMCMC">Evaluate MCMC</h2>

<p>We will use <code class="highlighter-rouge">Tracer</code>to evaluate the MCMC samples from our
three estimations. Load all three of the MCMC logs into the
<code class="highlighter-rouge">Tracer</code>window. The MCMC chains will not have converged
because they have not been run very long. Highlight all three files in
the upper left-hand viewer (Fig. [fig:tracer-files]) by right- or
command-clicking all three files.</p>

<p>[fig:tracer-files]</p>

<p><img src="figures/AddFiles.png" alt="" /></p>
<blockquote>
  <p>Highlight all three files for model
comparison.</p>
</blockquote>

<p>Once all three trace logs are loaded and highlighted, first look at the
estimated marginal likelihoods. You will notice that the Mk model, as
originally proposed by <a href="#Lewis2001">(Lewis 2001)</a> is improved by allowing any state
frequency heterogeneity at all. The discretized model and the Dirichlet
model both represent improvements, but are fairly close in likelihood
score to each other (Fig. [fig:tracer-llik]). Likely, we would need to
perform stepping stone model assessment to truly tell if the more
complicated model is statistically justified. This analysis is too
complicated and time-consuming for this tutorial period, but you will
find instructions below for performing the analysis.</p>

<p>[fig:tracer-llik]</p>

<p><img src="figures/likelihoods.png" alt="" /></p>
<blockquote>
  <p>Comparison of likelihood scores
for all three models.</p>
</blockquote>

<p>Click on the ‘Trace’ panel. In the lower left hand corner, you will
notice an option to color each trace by the file it came from. Choose
this option (you may need to expand the window slightly to see it). Next
to this option, you can also see an option to add a legend to your trace
window. The results of this coloring can be seen in
Fig. [fig:coltrace]. When the coloring is working, you will see that
the Mk model mixes quite well, but that mixing becomes worse as we relax
the assumption of equal state frequencies. This is because we are
greatly increasing model complexity. Therefore, we would need to run the
MCMC chains longer if we were to use these analyses in a paper.</p>

<p>[fig:coltrace]</p>

<p><img src="figures/colortrace.png" alt="" /></p>
<blockquote>
  <p>The Trace window. The traces are
colored by which version of the Mk model they correspond to.</p>
</blockquote>

<p>We are interested in two aspects of the posterior distribution. First,
all analyses correct for the biased sampling of variable characters
except for the simple analysis. Then, we expect the
tree_length variable to be greater for simple
than for the remaining analyses, because our data are enriched for
variation. Figure [fig:tracer_tree_length] shows that
tree_length is approximately 30% greater for
simple than for mk_simple, which are
identical except that mk_simple corrects for sampling
bias. To compare these densities, click the “Marginal Prob Distribution”
tab in the upper part of the window, highlight all of the loaded Trace
Files, then select tree_length from the list of Traces.</p>

<p>[fig:tracer_tree_length]</p>

<p><img src="figures/results/tracer_tree_length.png" alt="" /></p>
<blockquote>
  <p>Posterior tree
length estimates.</p>
</blockquote>

<p>Second, we are interested in characterizing the degree of heterogeneity
estimated by the beta-discretized model. If the data were distributed by
a single morphological rate matrix, then we would expect to see very
little variation among the different values in cats, and
very large values for the shape and scale parameters of the
discrete-beta distribution. For example, if alpha_ofbeta =
beta_ofbeta = 1000, then that would cause all discrete-beta
categories to have values approaching 0.5, which approximates a
symmetric Mk model.</p>

<p>[fig:tracer_cats]</p>

<p><img src="figures/results/cats.png" alt="" /></p>
<blockquote>
  <p>Posterior discretized state
frequencies for the discrete-beta model.</p>
</blockquote>

<p>[fig:tracer_alpha_beta]</p>

<p><img src="figures/results/alpha_beta.png" alt="" /></p>
<blockquote>
  <p>Posterior alpha and beta
parameters for the discrete-beta model.</p>
</blockquote>

<p>Figure [fig:tracer_cats] shows that the four discrete-beta state
frequencies do not all have the exact same value. In addition,
Figure [fig:tracer_alpha_beta] shows that the priors on the
discrete-beta distribution are small enough that we expect to see
variance among cat values. If the data contained no
information regarding the distribution of cat values, then
the posterior estimates for alpha_ofbeta and
beta_ofbeta would resemble the prior.</p>

<h2 id="summarizing-tree-estimates">Summarizing tree estimates</h2>

<p>The morphology trees estimated in [sec:dm_simple] and
[sec:dm_disc] are summarized using a majority rule consensus tree
(MRCT). Clades appearing in $p&gt;0.5$ of posterior samples are resolved in
the MRCT, while poorly support clades with $p \leq 0.5$ are shown as
unresolved polytomies. Poor phylogenetic resolution might be caused by
having too few phylogenetically informative characters, or it might be
due to conflicting signals for certain species relationships. Because
phylogenetic information is generated through model choice, let’s
compare our topological estimates across models.</p>

<p>[fig:mk_discretized_majrule]</p>

<p><img src="figures/results/mk_discretized_majrule_tre.png" alt="" /></p>
<blockquote>
  <p>Majority
rule consensus tree for the beta-discretized Mkv analysis.</p>
</blockquote>

<p>The MRCTs for the simple model with and without the +v correction are
very similar to that for the discretized-beta model
(Fig. [fig:mk_discretized_majrule]). Note that the scale bars for
branch lengths differ greatly, indicating that tree length estimates are
inflated without the +v correction, just as we saw when comparing the
posterior tree length densities. In general, it is important to assess
whether your results are sensitive to model assumptions, such as the
degree of model complexity, and any mechanistic assumptions that
motivate the model’s design. In this case, our tree estimate appears to
be robust to model complexity.</p>

<p>Version dated:</p>

<div class="footnotes">
  <ol>
    <li id="fn:1">
      <p>In section [subsub:Req-Software] we offer a recommendation for a
text editor. <a href="#fnref:1" class="reversefootnote">&#8617;</a></p>
    </li>
    <li id="fn:2">
      <p><a href="https://github.com/revbayes/revbayes_tutorial/tree/master/RB_DiscreteMorphology_Tutorial/scripts">https://github.com/revbayes/revbayes_tutorial/tree/master/RB_DiscreteMorphology_Tutorial/scripts</a> <a href="#fnref:2" class="reversefootnote">&#8617;</a></p>
    </li>
  </ol>
</div>


      <ol class="bibliography"><li><span id="Hoehna2016b">Höhna S., Landis M.J., Heath T.A., Boussau B., Lartillot N., Moore B.R., Huelsenbeck J.P., Ronquist F. 2016. RevBayes: Bayesian Phylogenetic Inference Using Graphical Models and an Interactive Model-Specification Language. Systematic Biology. 65:726–736.</span>

<a href="https://doi.org/10.1093/sysbio/syw021">10.1093/sysbio/syw021</a>

</li>
<li><span id="Phillips06">Phillips A.J. 2006. Homology assessment and molecular sequence alignment. Journal of Biomedical Informatics. 39:18–33.</span>

<a href="https://doi.org/10.1016/j.jbi.2005.11.005">10.1016/j.jbi.2005.11.005</a>

</li>
<li><span id="freudenstein05">Freudenstein J.V. 2005. Characters, States and Homology. Systematic Biology. 54:965.</span>

<a href="https://doi.org/10.1080/10635150500354654">10.1080/10635150500354654</a>

</li>
<li><span id="Lewis2001">Lewis P.O. 2001. A Likelihood Approach to Estimating Phylogeny from Discrete Morphological Character Data. Systematic Biology. 50:913–925.</span>

<a href="https://doi.org/10.1080/106351501753462876">10.1080/106351501753462876</a>

</li>
<li><span id="Jukes1969">Jukes T.H., Cantor C.R. 1969. Evolution of Protein Molecules. Mammalian Protein Metabolism. 3:21–132.</span>

<a href="https://doi.org/10.1016/B978-1-4832-3211-9.50009-7">10.1016/B978-1-4832-3211-9.50009-7</a>

</li>
<li><span id="Wright2014">Wright A.M., Hillis D.M. 2014. Bayesian analysis using a simple likelihood model outperforms parsimony for estimation of phylogeny from discrete morphological data. PLoS One. 9:e109210.</span>

<a href="https://doi.org/10.1371/journal.pone.0109210">10.1371/journal.pone.0109210</a>

</li>
<li><span id="allman08b">Allman E.S., Rhodes J.A. 2008. Identifying evolutionary trees and substitution parameters for the general Markov model with invariable sites. Mathematical Biosciences. 211:18–33.</span>

</li>
<li><span id="abella2011">Abella J., Montoya P., Morales J. 2011. Una nueva especie de \textitAgriarctos (Ailuropodinae, Ursidae, Carnivora) en la localidad de Nombrevilla 2 (Zaragoza, España). Estudios Geológicos. 67:187–191.</span>

</li>
<li><span id="Pagel2004">Pagel M., Meade A. 2004. A Phylogenetic Mixture Model for Detecting Pattern-Heterogeneity in Gene Sequence or Character-State Data. Systematic Biology. 53:571–581.</span>

<a href="https://doi.org/10.1080/10635150490468675">10.1080/10635150490468675</a>

</li>
<li><span id="Wright2016">Wright A.M., Lloyd G.T., Hillis D.M. 2016. Modeling Character Change Heterogeneity in Phylogenetic Analyses of Morphology through the Use of Priors. Systematic Biology. 65:602–611.</span>

<a href="https://doi.org/10.1093/sysbio/syv122">10.1093/sysbio/syv122</a>

</li>
<li><span id="Lartillot2004">Lartillot N., Philippe H. 2004. A Bayesian mixture model for across-site heterogeneities in the amino-acid replacement process. Molecular Biology and Evolution. 21:1095–1109.</span>

</li></ol>

<script type="text/javascript">
var _ol = document.querySelectorAll('ol');
for (var i = 0, elem_ol; elem_ol = _ol[i]; i++) {
	if ( elem_ol.classList == "bibliography" ) {
		var _li = elem_ol.getElementsByTagName("li");
		//for (var j = 0, elem_li; elem_li = _li[j]; j++)
		//{
		//	elem_li.innerHTML = elem_li.innerHTML.replace(/(https?:\/\/)([^\s<]+)/,"<a href=\"$1$2\">$2");
		//}
		if(_li.length > 0)
			elem_ol.outerHTML = "<h2>References</h2>"+elem_ol.outerHTML
	}
}
</script>
      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a> | <a href="/revbayes_tutorials/license">License</a> | <a href="/revbayes_tutorials/citation">Cite RevBayes</a> | <a href="https://groups.google.com/forum/#!forum/revbayes-users">Users' Forum</a>
    </div>
  </div>
  </div>
</footer>

    </div>
    <script src="/revbayes_tutorials/assets/js/vendor/jquery.min.js"></script>
<script src="/revbayes_tutorials/assets/js/vendor/FileSaver.min.js"></script>
<script src="/revbayes_tutorials/assets/js/vendor/jszip.min.js"></script>
<script src="/revbayes_tutorials/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "Rev highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "Rev highlighter-rouge";
//   }
//   
// });
</script>

<script src="/revbayes_tutorials/assets/js/base.js"></script>

<script type="text/javascript"
    src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
  });
</script>

  </body>
</html>
