<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="https://revbayes.github.io/revbayes-site/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/revbayes-site/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/revbayes-site/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/revbayes-site/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/revbayes-site/assets/css/main.css" />
    <title>RevBayes: Phylogenetic Inference using ‘RevBayes‘</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <a href="/revbayes-site/" class="pull-left">
        <img class="navbar-logo" src="/revbayes-site/assets/img/aquabayes-desaturated.png" alt="RevBayes Logo" />
      </a>
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar" align="right"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
        <li><a href="/revbayes-site/software/">Software</a></li>
        <li><a href="/revbayes-site/tutorials/">Tutorials</a></li>
        <li><a href="/revbayes-site/workshops/">Workshops</a></li>
        <li><a href="/revbayes-site/developer/">Developer</a></li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <div class="row">
	<h1 class="maintitle">Phylogenetic Inference using ‘RevBayes‘</h1>
	<h3 class="subtitle">Model selection using Bayes factors</h3>
	<h4 class="maintitle">Sebastian H&#246;hna, Michael J Landis, Tracy A Heath and Brian R Moore</h4>
</div>


<blockquote class="overview no-print" id="overview">
  <h2>Overview</h2>
  <div class="row">
    <div class="col-md-9">
        <strong>Prerequisites</strong>
        
          <ul id="prerequisites">
          
            
            
            <li><a href="/revbayes-site/tutorials/intro/">Statistical Inference Using RevBayes</a></li>
          
          </ul>
        
    </div>
  </div>
</blockquote>

<blockquote class="tutorial_files no-print" id="tutorial_files">
  <h2>Data files and scripts</h2>
  
    <div class="row" id="script_row">
      <div class="col-md-9">
          <strong>Scripts</strong>
          <ul id="scripts"></ul>
        </div>
    </div>
</blockquote>
<h1 id="overview">Overview</h1>

<p>This tutorial provides the third protocol from our recent publication
<a href="#Hoehna2017a">(Höhna et al. 2017)</a>. The first protocol is described in the <a href="https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_CTMC_Tutorial/RB_CTMC_Tutorial.pdf">Substitution
model
tutorial</a>
and the second protocol is described in the <a href="https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_Partition_Tutorial/RB_Partition_Tutorial.pdf">Partitioned data analysis
tutorial</a>.</p>

<p>This tutorial demonstrates some general principles of Bayesian model
comparison, which is based on estimating the marginal likelihood of
competing models and then comparing their relative fit to the data using
Bayes factors. We consider the specific case of calculating Bayes
factors to select among different substitution models.</p>

<h2 id="requirements">Requirements</h2>

<p>We assume that you have read and hopefully completed the following
tutorials:</p>

<ul>
  <li>
    <p><a href="https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_Getting_Started/RB_Getting_Started.pdf">Getting
started</a></p>
  </li>
  <li>
    <p><a href="https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_Intro_Tutorial/RB_Intro_Tutorial.pdf">‘Rev‘
basics</a></p>
  </li>
  <li>
    <p><a href="https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_Rev_Tutorial/RB_Rev_Tutorial.pdf">‘Rev‘
syntax</a></p>
  </li>
  <li>
    <p><a href="https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_CTMC_Tutorial/RB_CTMC_Tutorial.pdf">Substitution
models</a></p>
  </li>
</ul>

<p>Note that the <a href="https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_Intro_Tutorial/RB_Intro_Tutorial.pdf">‘Rev‘ basics
tutorial</a>
introduces the basic syntax of ‘Rev‘ but does not cover any phylogenetic
models. We tried to keep this tutorial very basic and introduce all the
language concepts and theory on the way. You may only need the <a href="https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_Rev_Tutorial/RB_Rev_Tutorial.pdf">‘Rev‘
syntax
tutorial</a>
for a more in-depth discussion of concepts in ‘Rev‘.</p>

<h1 id="data-and-files">Data and files</h1>

<p>We provide the data file that we will use in this tutorial. Of course,
you may want to use your own dataset instead. In the ‘data‘ folder, you
will find the following file:</p>

<ul>
  <li>‘primates_and_galeopterus_cytb.nex‘: Alignment of the <em>cytochrome
b</em> subunit from 23 primates representing 14 of the 16 families
(<em>Indriidae</em> and <em>Callitrichidae</em> are missing).</li>
</ul>

<h1 id="introduction">Introduction</h1>

<p>For most sequence alignments, several (possibly many) substitution
models of varying complexity are plausible <em>a priori</em>. We therefore need
a way to objectively identify the model that balances estimation bias
and inflated error variance associated with under- and
over-parameterized models, respectively. Increasingly, model selection
is based on <em>Bayes factors</em> [<em>e.g.,</em>
@Suchard2001} (missing reference):
<script type="math/tex">\begin{aligned}
\label{BFeq1}
BF(M_0,M_1) = \frac{\mathbb{P}(M_0 \mid \mathbf X, \theta_0)}{\mathbb{P}(M_1 \mid \mathbf X, \theta_1)} \div \frac{\mathbb{P}(M_0)}{\mathbb{P}(M_1)}. \end{aligned}</script>
Note that interpreting Bayes factors involves some subjectivity. That
is, it is up to <em>you</em> to decide the degree of your belief in $M_0$
relative to $M_1$. Despite the absence of an absolutely objective
model-selection threshold, we can refer to the scale [outlined by
@Jeffreys1961] that provides a “rule-of-thumb” for interpreting these
measures (Table [bftable]).</p>

<p>l c c c &amp; &amp; &amp;\
Negative (supports $M_1$) &amp; $&lt;1$ &amp; $&lt;0$ &amp; $&lt;0$\
Barely worth mentioning &amp; $1$ to $3.2$ &amp; $0$ to $1.16$ &amp; $0$ to $0.5$\
Substantial &amp; $3.2$ to $10$ &amp; $1.16$ to $2.3$ &amp; $0.5$ to $1$\
Strong &amp; $10$ to $100$ &amp; $2.3$ to $4.6$ &amp; $1$ to $2$\
Decisive&amp; $&gt;100$ &amp; $&gt;4.6$ &amp; $&gt;2$\</p>

<p>Unfortunately, it is generally not possible to directly calculate the
posterior odds to prior odds ratios. However, we can further define the
posterior odds ratio as: <script type="math/tex">\begin{aligned}
\frac{\mathbb{P}(M_0 \mid \mathbf X)}{\mathbb{P}(M_1 \mid \mathbf X)} = \frac{\mathbb{P}(M_0)}{\mathbb{P}(M_1)} \frac{\mathbb{P}(\mathbf X \mid M_0)}{\mathbb{P}(\mathbf X \mid M_1)},\end{aligned}</script>
where $\mathbb{P}(\mathbf X \mid M_i)$ is the <em>marginal likelihood</em> of
the data (this may be familiar to you as the denominator of Bayes
Theorem, which is variously referred to as the <em>model evidence</em> or
<em>integrated likelihood</em>). Formally, the marginal likelihood is the
probability of the observed data ($\mathbf X$) under a given model
($M_i$) that is averaged over all possible values of the parameters of
the model ($\theta_i$) with respect to the prior density on $\theta_i$
<script type="math/tex">\begin{aligned}
\label{margeLike}
\mathbb{P}(\mathbf X \mid M_i) = \int \mathbb{P}(\mathbf X \mid \theta_i) \mathbb{P}(\theta_i)dt.\end{aligned}</script>
This makes it clear that more complex (parameter-rich) models are
penalized by virtue of the associated prior: each additional parameter
entails integration of the likelihood over the corresponding prior
density. If you refer back to equation [BFeq1], you can see that, with
very little algebra, the ratio of marginal likelihoods is equal to the
Bayes factor: <script type="math/tex">\begin{aligned}
\label{bfFormula}
BF(M_0,M_1) = \frac{\mathbb{P}(\mathbf X \mid M_0)}{\mathbb{P}(\mathbf X \mid M_1)} = \frac{\mathbb{P}(M_0 \mid \mathbf X, \theta_0)}{\mathbb{P}(M_1 \mid \mathbf X, \theta_1)} \div \frac{\mathbb{P}(M_0)}{\mathbb{P}(M_1)}. \end{aligned}</script>
Therefore, we can perform a Bayes factor comparison of two models by
calculating the marginal likelihood for each one. Alas, exact solutions
for calculating marginal likelihoods are not known for phylogenetic
models (see equation [margeLike]), thus we must resort to numerical
integration methods to estimate or approximate these values. In this
exercise, we will estimate the marginal likelihood for each partition
scheme using both the stepping-stone (missing reference) and path
sampling estimators (missing reference).</p>

<h2 id="substitution-models">Substitution Models</h2>

<p>The models we use here are equivalent to the models described in the
previous exercise on substitution models (continuous time Markov
models). To specify the model please consult the previous exercise.
Specifically, you will need to specify the following substitution
models:</p>

<ul>
  <li>
    <p>Jukes-Cantor (JC) substitution model <a href="#Jukes1969">(Jukes and Cantor 1969)</a></p>
  </li>
  <li>
    <p>Hasegawa-Kishino-Yano (HKY) substitution model <a href="#Hasegawa1985">(Hasegawa et al. 1985)</a></p>
  </li>
  <li>
    <p>General-Time-Reversible (GTR) substitution model <a href="#Tavare1986">(Tavaré 1986)</a></p>
  </li>
  <li>
    <p>Gamma (+G) model for among-site rate variation <a href="#Yang1994a">(Yang 1994)</a></p>
  </li>
  <li>
    <p>Invariable-sites (+I) model <a href="#Hasegawa1985">(Hasegawa et al. 1985)</a></p>
  </li>
</ul>

<h2 id="estimating-the-marginal-likelihood">Estimating the Marginal Likelihood</h2>

<p>We will estimate the marginal likelihood of a given model using a
‘stepping-stone’ (or ‘path-sampling’) algorithm. These algorithms are
similar to the familiar MCMC algorithms, which are intended to sample
from (and estimate) the joint posterior probability of the model
parameters. Stepping-stone algorithms are like a series of MCMC
simulations that iteratively sample from a specified number of
distributions that are discrete steps between the posterior and the
prior probability distributions. The basic idea is to estimate the
probability of the data for all points between the posterior and the
prior—effectively summing the probability of the data over the prior
probability of the parameters to estimate the marginal likelihood.
Technically, the steps correspond to a series of ‘powerPosteriors()‘,
where the likelihood is iteratively raised to a series of numbers
between 1 and 0 (Figure [fig:ss]). When the likelihood is raised to
the power of 1 (typically the first stepping stone), samples are drawn
from the (untransformed) posterior. By contrast, when the likelihood is
raised to the power of 0 (typically the last stepping stone), samples
are drawn from the prior. To perform a stepping-stone simulation, we
need to specify (1) the number of stepping stones (power posteriors)
that we will use to traverse the path between the posterior and the
prior (<em>e.g.,</em>we specify 50 or 100 stones),
(2) the spacing of the stones between the posterior and prior
(<em>e.g.,</em>we may specify that the stones are
distributed according to a beta distribution), (3) the number of samples
(and their thinning) to be drawn from each stepping stone, and (4) the
direction we will take (<em>i.e.,</em>from the
posterior to the prior or vice versa).</p>

<blockquote class="figure">
  <p><img src="figures/ss.png" alt="" /> 
Estimating marginal likelihoods using
stepping-stone simulation. Estimating the marginal likelihood involves
integrating the likelihood of the data over the entire prior probability
density for the model parameters.MCMC algorithms target the posterior
probability density, which is typically concentrated in a small region
of the prior probability density (A).Accordingly, standard MCMC
simulation cannot provide unbiased estimates of the marginal likelihood
because it will typically fail to explore most of the prior density.(B)
Stepping-stone algorithms estimate the marginal likelihood by means of a
series of MCMC-like simulations, where the likelihood is iteratively
raised to a series of powers, effectively forcing the simulation to more
fully explore the prior density of the model parameters.Here, six
uniformly spaced stones span the posterior, where the power posterior is
$\beta=6/6=1$, to the prior, where the power posterior is $\beta=0/6=0$.</p>
</blockquote>

<p>This method computes a vector of powers from a beta distribution, then
executes an MCMC run for each power step while raising the likelihood to
that power. In this implementation, the vector of powers starts with 1,
sampling the likelihood close to the posterior and incrementally
sampling closer and closer to the prior as the power decreases.</p>

<p>Just to be safe, it is better to clear the workspace (if you did not
just restart ‘RevBayes‘):</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>clear()
</code></pre></div></div>

<p>Now set up the model as in the previous exercise. You should start with
the simple Jukes-Cantor substitution model. Setting up the model
requires:</p>

<ol>
  <li>
    <p>Loading the data and retrieving useful variables about it
(<em>e.g.,</em>number of sequences and
taxon names).</p>
  </li>
  <li>
    <p>Specifying the instantaneous-rate matrix of the substitution model.</p>
  </li>
  <li>
    <p>Specifying the tree model including branch-length variables.</p>
  </li>
  <li>
    <p>Creating a random variable for the sequences that evolved under
the ‘PhyloCTMC‘.</p>
  </li>
  <li>
    <p>Clamping the data.</p>
  </li>
  <li>
    <p>Creating a model object.</p>
  </li>
  <li>
    <p>Specifying the moves for parameter updates.</p>
  </li>
</ol>

<p>The following procedure for estimating marginal likelihoods is valid for
any model in ‘RevBayes‘. You will need to repeat this later for other
models. First, we create the variable containing the power-posterior
analysis. This requires that we provide a model and vector of moves, as
well as an output file name. The ‘cats‘ argument sets the number of
stepping stones.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pow_p = powerPosterior(mymodel, moves, monitors, "output/model1.out", cats=50) 
</code></pre></div></div>

<p>We can start the power-posterior analysis by first burning in the chain
and and discarding the first 10000 states. This will help ensure that
analysis starts from a region of high posterior probability, rather than
from some random point.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pow_p.burnin(generations=10000,tuningInterval=1000)
</code></pre></div></div>

<p>Now execute the run with the ‘.run()‘ function:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pow_p.run(generations=1000)  
</code></pre></div></div>

<p>Once the power posteriors have been saved to file, create a stepping
stone sampler. This function can read any file of power posteriors and
compute the marginal likelihood using stepping-stone sampling.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ss = steppingStoneSampler(file="output/model1.out", powerColumnName="power", likelihoodColumnName="likelihood")
</code></pre></div></div>

<p>These commands will execute a stepping-stone simulation with 50 stepping
stones, sampling 1000 states from each step. Compute the marginal
likelihood under stepping-stone sampling using the member function
‘marginal()‘ of the ‘ss‘ variable and record the value in Table
[tab:ml_cytb].</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ss.marginal() 
</code></pre></div></div>

<p>Path sampling is an alternative to stepping-stone sampling and also
takes the same power posteriors as input.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ps = pathSampler(file="output/model1.out", powerColumnName="power", likelihoodColumnName="likelihood")
</code></pre></div></div>

<p>Compute the marginal likelihood under stepping-stone sampling using the
member function ‘marginal()‘ of the ‘ps‘ variable and record the value
in Table [tab:ml_cytb].</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ps.marginal() 
</code></pre></div></div>

<p>\
As an example we provide the file
<strong>RevBayes_scripts/marginalLikelihood_JukesCantor.Rev</strong>.</p>

<p>We have kept this description of how to use stepping-stone-sampling and
path-sampling very generic and did not provide the information about the
model here. Our main motivation is to show that the marginal likelihood
estimation algorithms are independent of the model. Thus, you can apply
these algorithms to any model, <em>e.g.,</em>relaxed
clock models and birth-death models, as well.</p>

<h2 id="exercises">Exercises</h2>

<ul>
  <li>
    <p>Compute the marginal likelihoods of the <em>cytb</em> alignment for the
following substitution models:</p>

    <ul>
      <li>
        <p>Jukes-Cantor (JC) substitution model</p>
      </li>
      <li>
        <p>Hasegawa-Kishino-Yano (HKY) substitution model</p>
      </li>
      <li>
        <p>General-Time-Reversible (GTR) substitution model</p>
      </li>
      <li>
        <p>GTR with gamma distributed-rate model (GTR+G)</p>
      </li>
      <li>
        <p>GTR with invariable-sites model (GTR+I)</p>
      </li>
      <li>
        <p>GTR+I+G model</p>
      </li>
    </ul>
  </li>
  <li>
    <p>Enter the marginal likelihood estimate for each model in the
corresponding cell of Table [tab:ml_cytb].</p>
  </li>
  <li>
    <p>Which is the best fitting substitution model?</p>
  </li>
</ul>

<p>l c c c c &amp; &amp;\
&amp; &amp; &amp; &amp;\
JC ($M_1$) &amp;</p>

<p>&amp; &amp;</p>

<p>&amp;\
HKY ($M_2$) &amp;</p>

<p>&amp; &amp;</p>

<p>&amp;\
GTR ($M_3$) &amp;</p>

<p>&amp; &amp;</p>

<p>&amp;\
GTR+$\Gamma$ ($M_4$) &amp;</p>

<p>&amp; &amp;</p>

<p>&amp;\
GTR+I ($M_5$) &amp;</p>

<p>&amp; &amp;</p>

<p>&amp;\
GTR+$\Gamma$+I ($M_6$) &amp;</p>

<p>&amp; &amp;</p>

<p>&amp;\
\</p>

<p>[tab:ml_cytb]</p>

<h1 id="computing-bayes-factors-and-model-selection">Computing Bayes Factors and Model Selection</h1>

<p>Now that we have estimates of the marginal likelihood for each of our
the candidate substitution models, we can evaluate their relative fit to
the datasets using Bayes factors. Phylogenetic programs log-transform
the likelihood values to avoid
<a href="http://en.wikipedia.org/wiki/Arithmetic_underflow">underflow</a>:
multiplying likelihoods (numbers $&lt; 1$) generates numbers that are too
small to be held in computer memory. Accordingly, we need to use a
different form of equation [bfFormula] to calculate the ln-Bayes
factor (we will denote this value $\mathcal{K}$): <script type="math/tex">\begin{aligned}
\label{LNbfFormula}
\mathcal{K}=\ln[BF(M_0,M_1)] = \ln[\mathbb{P}(\mathbf X \mid M_0)]-\ln[\mathbb{P}(\mathbf X \mid M_1)],\end{aligned}</script>
where $\ln[\mathbb{P}(\mathbf X \mid M_0)]$ is the <em>marginal lnL</em>
estimate for model $M_0$. The value resulting from equation
[LNbfFormula] can be converted to a raw Bayes factor by simply taking
the exponent of $\cal{K}$ <script type="math/tex">\begin{aligned}
\label{LNbfFormula2}
BF(M_0,M_1) = e^{\cal{K}}.\end{aligned}</script> Alternatively, you can
directly interpret the strength of evidence in favor of $M_0$ in log
space by comparing the values of $\cal{K}$ to the appropriate scale
(Table [bftable], second column). In this case, we evaluate $\cal{K}$
in favor of model $M_0$ against model $M_1$ so that:</p>

<hr />
<p>if $\mathcal{K} &gt; 1$, model $M_0$ is preferred
  if $\mathcal{K} &lt; -1$, model $M_1$ is preferred.
  ————————————————–</p>

<p>Thus, values of $\mathcal{K}$ around 0 indicate that there is no
preference for either model.</p>

<p>Using the values you entered in Table [tab:ml_cytb] and equation
[LNbfFormula], calculate the ln-Bayes factors (using $\mathcal{K}$)
for each model comparison. Enter your answers in Table [bfTable2]
using the stepping-stone and the path-sampling estimates of the marginal
log-likelihoods.</p>

<p>[bfTable2]</p>

<h1 id="for-your-consideration">For your consideration…</h1>

<p>In this tutorial you have learned how to use ‘RevBayes‘ to assess the
<em>relative</em> fit of a pool of candidate substitution models to a given
sequence alignment. Typically, once we have identified the “best”
substitution model for our alignment, we would then proceed to use this
model for inference. Technically, this is a decision to condition our
inferences on the selected model, which explicitly assumes that it
provides a reasonable description of the process that gave rise to our
data. However, there are several additional issues to consider before
proceeding along these lines, which we briefly mention below.</p>

<h2 id="accommodating-model-uncertainty">Accommodating Model Uncertainty</h2>

<p>In some or many situations the number of possible models to compare is
large, <em>e.g.,</em>choosing all possible
combinations of substitution models <a href="#Huelsenbeck2004">(Huelsenbeck et al. 2004)</a>. Furthermore,
imagine, for example, that there are several (possibly many) alternative
models that provide a similarly good fit to our given dataset. In such
scenarios, conditioning inference on <em>any</em> single model (even the
‘best’) ignores uncertainty in the chosen model, which can cause
estimates to be biased. This is the issue of <em>model uncertainty</em>. The
Bayesian framework provides a natural approach for accommodating model
uncertainty by means of <em>model averaging</em>; we simply adopt the
perspective that models (like standard parameters) are random variables,
and integrate the inference over the distribution of candidate models.
We will demonstrate how to accommodate model uncertainty using
‘RevBayes‘ in a separate tutorial, RB_ModelAveraging_Tutorial.</p>

<h2 id="assessing-model-adequacy">Assessing Model Adequacy</h2>

<p>In this tutorial, we used Bayes factors to assess the fit of various
substitution models to our sequence data, effectively establishing the
<em>relative</em> rank of the candidate models. Even if we have successfully
identified the very best model from the pool of candidates, however, the
preferred model may nevertheless be woefully inadequate in an <em>absolute</em>
sense. For this reason, it is important to consider <em>model adequacy</em>:
whether a given model provides a reasonable description of the process
that gave rise to our sequence data. We can assess the absolute fit of a
model to a given dataset using <em>posterior predictive simulation</em>. This
approach is based on the following premise: if the candidate model
provides a reasonable description of the process that gave rise to our
dataset, then we should be able to generate data under this model that
resemble our observed data. We will demonstrate how to assess model
adequacy using ‘RevBayes‘ in a separate tutorial,
RB_ModelAdequacy_Tutorial.</p>

<p>Version dated:</p>

<ol class="bibliography"><li><span id="Hoehna2017a">Höhna S., Landis M.J., Heath T.A. 2017. Phylogenetic Inference using RevBayes. Current Protocols in Bioinformatics.</span>

<a href="https://doi.org/10.1002/cpbi.22">10.1002/cpbi.22</a>

</li>
<li><span id="Jukes1969">Jukes T.H., Cantor C.R. 1969. Evolution of Protein Molecules. Mammalian Protein Metabolism. 3:21–132.</span>

<a href="https://doi.org/10.1016/B978-1-4832-3211-9.50009-7">10.1016/B978-1-4832-3211-9.50009-7</a>

</li>
<li><span id="Hasegawa1985">Hasegawa M., Kishino H., Yano T. 1985. Dating of the Human-Ape Splitting by a molecular Clock of Mitochondrial DNA. Journal of Molecular Evolution. 22:160–174.</span>

<a href="https://doi.org/10.1007/BF02101694">10.1007/BF02101694</a>

</li>
<li><span id="Tavare1986">Tavaré S. 1986. Some Probabilistic and Statistical Problems in the Analysis of DNA Sequences. Some Mathematical Questions in Biology: DNA Sequence Analysis. 17:57–86.</span>

</li>
<li><span id="Yang1994a">Yang Z. 1994. Maximum Likelihood Phylogenetic Estimation from DNA Sequences with Variable Rates Over Sites: Approximate Methods. Journal of Molecular Evolution. 39:306–314.</span>

<a href="https://doi.org/10.1007/BF00160154">10.1007/BF00160154</a>

</li>
<li><span id="Huelsenbeck2004">Huelsenbeck J.P., Larget B., Alfaro M.E. 2004. Bayesian Phylogenetic Model Selection using Reversible Jump Markov Chain Monte Carlo. Molecular Biology and Evolution. 21:1123.</span>

</li></ol>

<script type="text/javascript">
var _ol = document.querySelectorAll('ol');
for (var i = 0, elem_ol; elem_ol = _ol[i]; i++) {
	if ( elem_ol.classList == "bibliography" ) {
		var _li = elem_ol.getElementsByTagName("li");
		//for (var j = 0, elem_li; elem_li = _li[j]; j++)
		//{
		//	elem_li.innerHTML = elem_li.innerHTML.replace(/(https?:\/\/)([^\s<]+)/,"<a href=\"$1$2\">$2");
		//}
		if(_li.length > 0)
			elem_ol.outerHTML = "<h2>References</h2>"+elem_ol.outerHTML
	}
}
</script>

      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a> | <a href="/revbayes-site/license">License</a> | <a href="/revbayes-site/citation">Cite RevBayes</a> | <a href="https://groups.google.com/forum/#!forum/revbayes-users">Users' Forum</a>
    </div>
  </div>
  </div>
</footer>

    </div>
    <script src="/revbayes-site/assets/js/vendor/jquery.min.js"></script>
<script src="/revbayes-site/assets/js/vendor/FileSaver.min.js"></script>
<script src="/revbayes-site/assets/js/vendor/jszip.min.js"></script>
<script src="/revbayes-site/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "Rev highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "Rev highlighter-rouge";
//   }
//   
// });
</script>

<script src="/revbayes-site/assets/js/base.js"></script>

<script type="text/javascript"
    src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
  });
</script>

  </body>
</html>
